# Proyecto: Predicción de Lluvia - Colombia

## Limpieza de datos

### Importación de Librerias

In [3]:
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
import re
import os
import plotly.express as pex
from plotly.subplots import make_subplots
import plotly.graph_objects as go

*Sistema de archivos*

    ./
    
    ./Proyecto/
        ./Proyecto/Proyecto_bootcamp_cuaderno.ipynb

    ./Datasets/
        ./Datasets/...... .csv
    
    ./CleanDatasets/
        ./CleanDatasets/.....  .parquet


### Primera exploración

Se importa el primer documento previo a la limpieza. Se inicia con el archivo de menor tamaño. Todos los datos corresponden a precipitación y tienen la misma unidad de medida.

In [3]:
df = pd.read_csv('../Datasets/Precipitaci_n_20241016(vaupes).csv',
                 sep=',',
                 #nrows=10
                 )
df

CodigoEstacion  CodigoSensor        FechaObservacion  ValorObservado  \
0           42077020           240  11/01/2023 07:30:00 PM             0.2   
1           42077020           240  11/01/2023 11:10:00 PM             0.0   
2           42077020           240  11/01/2023 03:10:00 PM             0.0   
3           42077020           240  11/01/2023 09:30:00 PM             0.0   
4           42077020           240  11/01/2023 05:40:00 PM             0.0   
...              ...           ...                     ...             ...   
9147        42077020           240  06/22/2024 02:20:00 AM             0.0   
9148        42077020           240  06/22/2024 01:30:00 AM             0.0   
9149        42077020           240  06/22/2024 01:40:00 AM             0.0   
9150        42077020           240  06/22/2024 01:20:00 AM             0.0   
9151        42077020           240  06/22/2024 02:00:00 AM             0.0   

     NombreEstacion Departamento Municipio ZonaHidrografica  Latitud  \
0              MITU       VAUPES      MITÚ           VAUPES     1.26   
1              MITU       VAUPES      MITÚ           VAUPES     1.26   
2              MITU       VAUPES      MITÚ           VAUPES     1.26   
3              MITU       VAUPES      MITÚ           VAUPES     1.26   
4              MITU       VAUPES      MITÚ           VAUPES     1.26   
...             ...          ...       ...              ...      ...   
9147           MITU       VAUPES      MITÚ           VAUPES     1.26   
9148           MITU       VAUPES      MITÚ           VAUPES     1.26   
9149           MITU       VAUPES      MITÚ           VAUPES     1.26   
9150           MITU       VAUPES      MITÚ           VAUPES     1.26   
9151           MITU       VAUPES      MITÚ           VAUPES     1.26   

      Longitud DescripcionSensor UnidadMedida  
0       -70.24     PRECIPITACIÓN           mm  
1       -70.24     PRECIPITACIÓN           mm  
2       -70.24     PRECIPITACIÓN           mm  
3       -70.24     PRECIPITACIÓN           mm  
4       -70.24     PRECIPITACIÓN           mm  
...        ...               ...          ...  
9147    -70.24     PRECIPITACIÓN           mm  
9148    -70.24     PRECIPITACIÓN           mm  
9149    -70.24     PRECIPITACIÓN           mm  
9150    -70.24     PRECIPITACIÓN           mm  
9151    -70.24     PRECIPITACIÓN           mm  

[9152 rows x 12 columns]

En la siguiente celda se imprime un pequeño resumen. Es importante resaltar que no hay datos nulos y la necesidad de cambiar el formato de cada columna por uno mas adecuado que reduzca el tamaño del dataframe.

In [4]:
size_0=df.memory_usage(deep=True).sum()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9152 entries, 0 to 9151
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CodigoEstacion     9152 non-null   int64  
 1   CodigoSensor       9152 non-null   int64  
 2   FechaObservacion   9152 non-null   object 
 3   ValorObservado     9152 non-null   float64
 4   NombreEstacion     9152 non-null   object 
 5   Departamento       9152 non-null   object 
 6   Municipio          9152 non-null   object 
 7   ZonaHidrografica   9152 non-null   object 
 8   Latitud            9152 non-null   float64
 9   Longitud           9152 non-null   float64
 10  DescripcionSensor  9152 non-null   object 
 11  UnidadMedida       9152 non-null   object 
dtypes: float64(3), int64(2), object(7)
memory usage: 858.1+ KB


### Vaupes

Se importa nuevamente el archivo csv, seleccionando las columnas relevantes y eligiendo un formato adeacuado para cada columna.

In [5]:
df = pd.read_csv('../Datasets/Precipitaci_n_20241016(vaupes).csv',
                 sep=',',
                 usecols= [0,1,2,3,4,5,6,7,8,9],
                 dtype={0:'category',1:'category',2:'str',3:'float32',4:'category',5:'category',6:'category',7:'category',8:'category',9:'category'}
                 )
df

CodigoEstacion CodigoSensor        FechaObservacion  ValorObservado  \
0        0042077020         0240  11/01/2023 07:30:00 PM             0.2   
1        0042077020         0240  11/01/2023 11:10:00 PM             0.0   
2        0042077020         0240  11/01/2023 03:10:00 PM             0.0   
3        0042077020         0240  11/01/2023 09:30:00 PM             0.0   
4        0042077020         0240  11/01/2023 05:40:00 PM             0.0   
...             ...          ...                     ...             ...   
9147     0042077020         0240  06/22/2024 02:20:00 AM             0.0   
9148     0042077020         0240  06/22/2024 01:30:00 AM             0.0   
9149     0042077020         0240  06/22/2024 01:40:00 AM             0.0   
9150     0042077020         0240  06/22/2024 01:20:00 AM             0.0   
9151     0042077020         0240  06/22/2024 02:00:00 AM             0.0   

     NombreEstacion Departamento Municipio ZonaHidrografica Latitud Longitud  
0              MITU       VAUPES      MITÚ           VAUPES    1.26   -70.24  
1              MITU       VAUPES      MITÚ           VAUPES    1.26   -70.24  
2              MITU       VAUPES      MITÚ           VAUPES    1.26   -70.24  
3              MITU       VAUPES      MITÚ           VAUPES    1.26   -70.24  
4              MITU       VAUPES      MITÚ           VAUPES    1.26   -70.24  
...             ...          ...       ...              ...     ...      ...  
9147           MITU       VAUPES      MITÚ           VAUPES    1.26   -70.24  
9148           MITU       VAUPES      MITÚ           VAUPES    1.26   -70.24  
9149           MITU       VAUPES      MITÚ           VAUPES    1.26   -70.24  
9150           MITU       VAUPES      MITÚ           VAUPES    1.26   -70.24  
9151           MITU       VAUPES      MITÚ           VAUPES    1.26   -70.24  

[9152 rows x 10 columns]

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9152 entries, 0 to 9151
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   CodigoEstacion    9152 non-null   category
 1   CodigoSensor      9152 non-null   category
 2   FechaObservacion  9152 non-null   object  
 3   ValorObservado    9152 non-null   float32 
 4   NombreEstacion    9152 non-null   category
 5   Departamento      9152 non-null   category
 6   Municipio         9152 non-null   category
 7   ZonaHidrografica  9152 non-null   category
 8   Latitud           9152 non-null   category
 9   Longitud          9152 non-null   category
dtypes: category(8), float32(1), object(1)
memory usage: 178.9+ KB


Se muestra una comparación entre el tamaño del dataframe tras el cambio de formato a las columnas

In [7]:
size_f=df.memory_usage(deep=True).sum()
print(size_f/size_0)

0.1759772734217371


**Formato de fecha**

In [8]:
df['FechaObservacion'] = pd.to_datetime(df['FechaObservacion'], format='%m/%d/%Y %H:%M:%S %p')

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9152 entries, 0 to 9151
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   CodigoEstacion    9152 non-null   category      
 1   CodigoSensor      9152 non-null   category      
 2   FechaObservacion  9152 non-null   datetime64[ns]
 3   ValorObservado    9152 non-null   float32       
 4   NombreEstacion    9152 non-null   category      
 5   Departamento      9152 non-null   category      
 6   Municipio         9152 non-null   category      
 7   ZonaHidrografica  9152 non-null   category      
 8   Latitud           9152 non-null   category      
 9   Longitud          9152 non-null   category      
dtypes: category(8), datetime64[ns](1), float32(1)
memory usage: 178.9 KB


In [10]:
df.to_parquet('../CleanDatasets/rain_vaupes.parquet')

In [11]:
namesFiles = os.listdir('../Datasets')
namesFiles = [re.findall(r'\((\w*)\)',x)[0] for x in namesFiles]
print(namesFiles)

['amazonas', 'antioquia', 'arauca', 'atlantico1', 'atlantico2', 'bogota1', 'bogota2', 'bogota3', 'bolivar1', 'bolivar2', 'boyaca', 'caldas', 'caqueta1', 'caqueta2', 'casanare', 'cauca', 'cesar', 'choco', 'choco1', 'cordoba1', 'cordoba2', 'cundinamarca', 'guainia', 'guajira', 'guaviare', 'huila', 'magdalena', 'meta', 'narino', 'narino1', 'nill', 'norte_de_santander', 'putumayo', 'quindio', 'risaralda', 'sanandres1', 'sanandres2', 'sanandres3', 'sanandres4', 'santander', 'sucre', 'tolima', 'valle_del_cauca', 'vaupes', 'vaupes1', 'vichada']


### Caquetá

Tener en cuenta que la información sobre caquetá está almacenada en dos archivos

#### Caquetá 2

In [12]:
nameFile = '../Datasets/Precipitaci_n_20241016('+ namesFiles[13] +').csv'
nameFile

'../Datasets/Precipitaci_n_20241016(caqueta2).csv'

In [13]:
df_caqueta2 = pd.read_csv(nameFile,
                 sep=',',
                 usecols= [0,1,2,3,4,5,6,7,8,9],
                 dtype={0:'category',1:'category',2:'str',3:'float32',4:'category',5:'category',6:'category',7:'category',8:'category',9:'category'}
                 )
df_caqueta2['FechaObservacion'] = pd.to_datetime(df_caqueta2['FechaObservacion'], format='%m/%d/%Y %H:%M:%S %p')
df_caqueta2

CodigoEstacion CodigoSensor    FechaObservacion  ValorObservado  \
0         0044055010         0240 2023-09-13 01:50:00             0.0   
1         0044035501         0240 2023-09-18 06:30:00             0.0   
2         0044035501         0240 2023-09-19 07:50:00             0.0   
3         0044035501         0240 2023-09-19 07:30:00             0.0   
4         0044035501         0240 2023-09-18 09:20:00             0.0   
...              ...          ...                 ...             ...   
14541     0044055010         0240 2024-06-12 02:30:00             0.0   
14542     0044055010         0240 2024-06-12 06:30:00             0.0   
14543     0044055010         0240 2024-06-12 12:50:00             0.0   
14544     0044055010         0240 2024-06-12 02:10:00             0.0   
14545     0044055010         0240 2024-06-12 07:40:00             0.0   

      NombreEstacion Departamento  Municipio ZonaHidrografica     Latitud  \
0      TRES ESQUINAS      CAQUETÁ     SOLANO          CAQUETÁ      0.7375   
1             PAUJIL      CAQUETÁ  EL PAUJIL          CAQUETÁ  1.57381944   
2             PAUJIL      CAQUETÁ  EL PAUJIL          CAQUETÁ  1.57381944   
3             PAUJIL      CAQUETÁ  EL PAUJIL          CAQUETÁ  1.57381944   
4             PAUJIL      CAQUETÁ  EL PAUJIL          CAQUETÁ  1.57381944   
...              ...          ...        ...              ...         ...   
14541  TRES ESQUINAS      CAQUETÁ     SOLANO          CAQUETÁ      0.7375   
14542  TRES ESQUINAS      CAQUETÁ     SOLANO          CAQUETÁ      0.7375   
14543  TRES ESQUINAS      CAQUETÁ     SOLANO          CAQUETÁ      0.7375   
14544  TRES ESQUINAS      CAQUETÁ     SOLANO          CAQUETÁ      0.7375   
14545  TRES ESQUINAS      CAQUETÁ     SOLANO          CAQUETÁ      0.7375   

           Longitud  
0      -75.23611111  
1          -75.3402  
2          -75.3402  
3          -75.3402  
4          -75.3402  
...             ...  
14541  -75.23611111  
14542  -75.23611111  
14543  -75.23611111  
14544  -75.23611111  
14545  -75.23611111  

[14546 rows x 10 columns]

La siguiente celda imprimirá la cantidad de valores de nulos

In [14]:
nulls = df_caqueta2.count(axis=1).sum()
size = df_caqueta2.shape[0]*df_caqueta2.shape[1]
print(size-nulls)

0


De nuevo se observa que el archivo tiene un tamaño menor y que los diferentes formatos son aplicados correctamente a cada columna

In [15]:
df_caqueta2.info()
#df_caqueta2.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14546 entries, 0 to 14545
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   CodigoEstacion    14546 non-null  category      
 1   CodigoSensor      14546 non-null  category      
 2   FechaObservacion  14546 non-null  datetime64[ns]
 3   ValorObservado    14546 non-null  float32       
 4   NombreEstacion    14546 non-null  category      
 5   Departamento      14546 non-null  category      
 6   Municipio         14546 non-null  category      
 7   ZonaHidrografica  14546 non-null  category      
 8   Latitud           14546 non-null  category      
 9   Longitud          14546 non-null  category      
dtypes: category(8), datetime64[ns](1), float32(1)
memory usage: 285.1 KB


### Todos los departamentos

In [16]:
print(namesFiles)
print(len(namesFiles))

['amazonas', 'antioquia', 'arauca', 'atlantico1', 'atlantico2', 'bogota1', 'bogota2', 'bogota3', 'bolivar1', 'bolivar2', 'boyaca', 'caldas', 'caqueta1', 'caqueta2', 'casanare', 'cauca', 'cesar', 'choco', 'choco1', 'cordoba1', 'cordoba2', 'cundinamarca', 'guainia', 'guajira', 'guaviare', 'huila', 'magdalena', 'meta', 'narino', 'narino1', 'nill', 'norte_de_santander', 'putumayo', 'quindio', 'risaralda', 'sanandres1', 'sanandres2', 'sanandres3', 'sanandres4', 'santander', 'sucre', 'tolima', 'valle_del_cauca', 'vaupes', 'vaupes1', 'vichada']
46


In [24]:
for name in namesFiles[:]:
    nameFile = '../Datasets/Precipitaci_n_20241016('+ name +').csv'
    df_dept = pd.read_csv(nameFile,
                 sep=',',
                 usecols= [0,1,2,3,4,5,6,7,8,9],
                 dtype={0:'category',1:'category',2:'str',3:'float32',4:'category',5:'category',6:'category',7:'category',8:'category',9:'category'}
                 )
    df_dept['FechaObservacion'] = pd.to_datetime(df_dept['FechaObservacion'], format='%m/%d/%Y %H:%M:%S %p')
    nulls = df_dept.count(axis=1).sum()
    size = df_dept.shape[0]*df_dept.shape[1]
    if nulls-size != 0:
        print(f'{nulls-size} nulls in {name}, parquet file can not created')
    else:
        df_dept.to_parquet(f'../CleanDatasets/rain_{name}.parquet')
        print(f'Not nulls in {name}, parquet file created')


Not nulls in amazonas, parquet file created
Not nulls in antioquia, parquet file created
Not nulls in arauca, parquet file created
Not nulls in atlantico1, parquet file created
Not nulls in atlantico2, parquet file created
Not nulls in bogota1, parquet file created
Not nulls in bogota2, parquet file created
Not nulls in bogota3, parquet file created
Not nulls in bolivar1, parquet file created
Not nulls in bolivar2, parquet file created
Not nulls in boyaca, parquet file created
Not nulls in caldas, parquet file created
Not nulls in caqueta1, parquet file created
Not nulls in caqueta2, parquet file created
Not nulls in casanare, parquet file created
Not nulls in cauca, parquet file created
Not nulls in cesar, parquet file created
Not nulls in choco, parquet file created
Not nulls in choco1, parquet file created
Not nulls in cordoba1, parquet file created
Not nulls in cordoba2, parquet file created
Not nulls in cundinamarca, parquet file created
Not nulls in guainia, parquet file created


## Consolidación de archivos

Se requiere un conjunto de datos que contiene la información de las estaciones de monitoreo.

A continuación se listan los archivos de la segunda carpeta que contiene los datasets:

    - Catálogo nacional de estaciones (Contine el listado de las estaciones)
    - Datos de estaciones IDEAM y terceros (Contiene mediciones de diversas estaciones propias y externas al IDEAM)
    - Datos hidrometeorológicos (Mediciones de temperatura)
    - Presión atmosférica  

In [25]:
namesFiles2 = os.listdir('../Datasets2')
namesFiles2 = [x for x in namesFiles2]
print(namesFiles2)

['Cat_logo_Nacional_de_Estaciones_del_IDEAM_20241017.csv', 'Datos_de_Estaciones_de_IDEAM_y_de_Terceros_20241017.csv', 'Datos_Hidrometeorol_gicos_Crudos_-_Red_de_Estaciones_IDEAM___Temperatura_20241017(0).csv', 'Datos_Hidrometeorol_gicos_Crudos_-_Red_de_Estaciones_IDEAM___Temperatura_20241017(1).csv', 'Datos_Hidrometeorol_gicos_Crudos_-_Red_de_Estaciones_IDEAM___Temperatura_20241017(2).csv', 'Datos_Hidrometeorol_gicos_Crudos_-_Red_de_Estaciones_IDEAM___Temperatura_20241017(3).csv', 'Datos_Hidrometeorol_gicos_Crudos_-_Red_de_Estaciones_IDEAM___Temperatura_20241017(4).csv', 'Presi_n_Atmosf_rica_20241017.csv']


### Limpieza *DATOS ESTACIONES IDEAM*

La ruta de acceso al archivo de Datos estaciones IDEAM es f'../Datasets2/{namesFiles2[1]}'

In [26]:
df_estaciones = pd.read_csv(f'../Datasets2/{namesFiles2[1]}',
                            sep=',',
                            nrows=3
)
df_estaciones

CodigoEstacion  CodigoSensor        FechaObservacion  ValorObservado  \
0      2120500204            69  10/15/2024 09:04:00 PM            15.8   
1      2120500204            70  10/15/2024 09:59:00 PM            15.5   
2      2120500204            69  10/15/2024 10:25:00 PM            15.7   

        NombreEstacion Departamento   Municipio ZonaHidrografica   Latitud  \
0  IDEAM PUENTE ARANDA       BOGOTÁ  BOGOTÁ D.C   ALTO MAGDALENA  4.621556   
1  IDEAM PUENTE ARANDA       BOGOTÁ  BOGOTÁ D.C   ALTO MAGDALENA  4.621556   
2  IDEAM PUENTE ARANDA       BOGOTÁ  BOGOTÁ D.C   ALTO MAGDALENA  4.621556   

    Longitud                  DescripcionSensor UnidadMedida  \
0 -74.104709  TEMPERATURA DEL AIRE MÁXIMA A 2 m           °C   
1 -74.104709  TEMPERATURA MÍNIMA DEL AIRE A 2 m           °C   
2 -74.104709  TEMPERATURA DEL AIRE MÁXIMA A 2 m           °C   

                                             Entidad  
0  INSTITUTO DE HIDROLOGIA METEOROLOGIA Y ESTUDIO...  
1  INSTITUTO DE HIDROLOGIA METEOROLOGIA Y ESTUDIO...  
2  INSTITUTO DE HIDROLOGIA METEOROLOGIA Y ESTUDIO...

In [27]:
df_estaciones = pd.read_csv(f'../Datasets2/{namesFiles2[1]}',
                            sep=',',
                            dtype={0:'category',1:'category',2:'str',3:'float32',4:'category',5:'category',6:'category',7:'category',8:'category',9:'category',10:'category',11:'category',12:'category'}
                            #nrows=3
)
df_estaciones.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 202642 entries, 0 to 202641
Data columns (total 13 columns):
 #   Column             Non-Null Count   Dtype   
---  ------             --------------   -----   
 0   CodigoEstacion     202642 non-null  category
 1   CodigoSensor       202642 non-null  category
 2   FechaObservacion   202642 non-null  object  
 3   ValorObservado     202642 non-null  float32 
 4   NombreEstacion     202642 non-null  category
 5   Departamento       202642 non-null  category
 6   Municipio          202642 non-null  category
 7   ZonaHidrografica   202642 non-null  category
 8   Latitud            202642 non-null  category
 9   Longitud           202642 non-null  category
 10  DescripcionSensor  202642 non-null  category
 11  UnidadMedida       202642 non-null  category
 12  Entidad            202642 non-null  category
dtypes: category(11), float32(1), object(1)
memory usage: 5.3+ MB


Se evidencia que no hay valores nulos. Hacemos una conversión a formato de fecha y almacenamos el archivo en formato .parquet

In [28]:
df_estaciones['FechaObservacion'] = pd.to_datetime(df_estaciones['FechaObservacion'], format='%m/%d/%Y %H:%M:%S %p')
df_estaciones.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 202642 entries, 0 to 202641
Data columns (total 13 columns):
 #   Column             Non-Null Count   Dtype         
---  ------             --------------   -----         
 0   CodigoEstacion     202642 non-null  category      
 1   CodigoSensor       202642 non-null  category      
 2   FechaObservacion   202642 non-null  datetime64[ns]
 3   ValorObservado     202642 non-null  float32       
 4   NombreEstacion     202642 non-null  category      
 5   Departamento       202642 non-null  category      
 6   Municipio          202642 non-null  category      
 7   ZonaHidrografica   202642 non-null  category      
 8   Latitud            202642 non-null  category      
 9   Longitud           202642 non-null  category      
 10  DescripcionSensor  202642 non-null  category      
 11  UnidadMedida       202642 non-null  category      
 12  Entidad            202642 non-null  category      
dtypes: category(11), datetime64[ns](1), float32(

In [29]:
df_estaciones.to_parquet('../CleanDatasets2/estaciones.parquet')

### Limpieza *DATOS PRESIÓN ATMOSFÉRICA*

In [30]:
namesFiles2

['Cat_logo_Nacional_de_Estaciones_del_IDEAM_20241017.csv',
 'Datos_de_Estaciones_de_IDEAM_y_de_Terceros_20241017.csv',
 'Datos_Hidrometeorol_gicos_Crudos_-_Red_de_Estaciones_IDEAM___Temperatura_20241017(0).csv',
 'Datos_Hidrometeorol_gicos_Crudos_-_Red_de_Estaciones_IDEAM___Temperatura_20241017(1).csv',
 'Datos_Hidrometeorol_gicos_Crudos_-_Red_de_Estaciones_IDEAM___Temperatura_20241017(2).csv',
 'Datos_Hidrometeorol_gicos_Crudos_-_Red_de_Estaciones_IDEAM___Temperatura_20241017(3).csv',
 'Datos_Hidrometeorol_gicos_Crudos_-_Red_de_Estaciones_IDEAM___Temperatura_20241017(4).csv',
 'Presi_n_Atmosf_rica_20241017.csv']

La ruta de acceso a los datos de presión atmosférica es: f '../Datasets2/{namesFiles2[-1]}']

In [31]:
df_presion = pd.read_csv(
                        f'../Datasets2/{namesFiles2[-1]}',
                        sep=',',
                        nrows=3
)
df_presion

CodigoEstacion  CodigoSensor        FechaObservacion  ValorObservado  \
0        36015020           255  10/03/2017 06:00:00 AM           992.5   
1        21195190           255  02/14/2014 05:00:00 AM           785.2   
2        21015050           255  09/27/2013 06:00:00 PM           805.3   

      NombreEstacion  Departamento       Municipio ZonaHidrografica   Latitud  \
0  EL DIAMANTE - AUT      CASANARE  PAZ DE ARIPORO             META  5.816194   
1        PASCA - AUT  CUNDINAMARCA           PASCA   ALTO MAGDALENA  4.310111   
2      PURACE  - AUT         HUILA     SAN AGUSTÍN   ALTO MAGDALENA  1.925917   

    Longitud         DescripcionSensor UnidadMedida  
0 -71.419833  Presión Atmosferica (1h)          HPa  
1 -74.311750  Presión Atmosferica (1h)          HPa  
2 -76.427556  Presión Atmosferica (1h)          HPa

In [32]:
df_presion = pd.read_csv(
                        f'../Datasets2/{namesFiles2[-1]}',
                        sep=',',
                        dtype={0:'category',1:'category',2:'str',3:'float32',4:'category',5:'category',6:'category',7:'category',8:'category',9:'category',10:'category',11:'category'}
)
df_presion.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22611008 entries, 0 to 22611007
Data columns (total 12 columns):
 #   Column             Dtype   
---  ------             -----   
 0   CodigoEstacion     category
 1   CodigoSensor       category
 2   FechaObservacion   object  
 3   ValorObservado     float32 
 4   NombreEstacion     category
 5   Departamento       category
 6   Municipio          category
 7   ZonaHidrografica   category
 8   Latitud            category
 9   Longitud           category
 10  DescripcionSensor  category
 11  UnidadMedida       category
dtypes: category(10), float32(1), object(1)
memory usage: 582.3+ MB


In [33]:
df_presion['FechaObservacion'] = pd.to_datetime(df_presion['FechaObservacion'],format='%m/%d/%Y %H:%M:%S %p')
df_presion

CodigoEstacion CodigoSensor    FechaObservacion  ValorObservado  \
0            0036015020         0255 2017-10-03 06:00:00      992.500000   
1            0021195190         0255 2014-02-14 05:00:00      785.200012   
2            0021015050         0255 2013-09-27 06:00:00      805.299988   
3            0021115010         0255 2005-11-28 10:00:00      958.500000   
4            0028035060         0255 2008-04-08 04:00:00      988.700012   
...                 ...          ...                 ...             ...   
22611003     0048015050         0258 2024-10-16 11:04:00     1000.500000   
22611004     0048015050         0258 2024-10-16 09:08:00     1004.099976   
22611005     0016015501         0258 2024-10-16 07:02:00      972.900024   
22611006     0048015040         0255 2024-10-16 10:40:00      997.400024   
22611007     0011175000         0255 2024-10-16 02:00:00     1006.000000   

                   NombreEstacion        Departamento       Municipio  \
0               EL DIAMANTE - AUT            CASANARE  PAZ DE ARIPORO   
1                     PASCA - AUT        CUNDINAMARCA           PASCA   
2                   PURACE  - AUT               HUILA     SAN AGUSTÍN   
3                DESIERTO TATACOA               HUILA      VILLAVIEJA   
4                 FEDEARROZ - AUT               CESAR      VALLEDUPAR   
...                           ...                 ...             ...   
22611003  AEROPUERTO VASQUEZ COBO            AMAZONAS         LETICIA   
22611004  AEROPUERTO VASQUEZ COBO            AMAZONAS         LETICIA   
22611005   AEROPUERTO CAMILO DAZA  NORTE DE SANTANDER          CÚCUTA   
22611006            PUERTO NARIÑO            AMAZONAS   PUERTO NARIÑO   
22611007                   ATRATO               CHOCO        CÉRTEGUI   

             ZonaHidrografica       Latitud      Longitud  \
0                        META   5.816194444  -71.41983333   
1              ALTO MAGDALENA   4.310111111     -74.31175   
2              ALTO MAGDALENA   1.925916667  -76.42755556   
3              ALTO MAGDALENA         3.234       -75.168   
4                       CESAR   10.46361111  -73.24805556   
...                       ...           ...           ...   
22611003  AMAZONAS - DIRECTOS  -4.193861111  -69.94091667   
22611004  AMAZONAS - DIRECTOS  -4.193861111  -69.94091667   
22611005            CATATUMBO       7.93028     -72.50917   
22611006  AMAZONAS - DIRECTOS  -3.780305556  -70.36263889   
22611007      ATRATO - DARIÉN      5.585776    -76.649978   

                   DescripcionSensor UnidadMedida  
0           Presión Atmosferica (1h)          HPa  
1           Presión Atmosferica (1h)          HPa  
2           Presión Atmosferica (1h)          HPa  
3           Presión Atmosferica (1h)          HPa  
4           Presión Atmosferica (1h)          HPa  
...                              ...          ...  
22611003  GPRS - PRESIÓN ATMOSFÉRICA          hPA  
22611004  GPRS - PRESIÓN ATMOSFÉRICA          hPA  
22611005  GPRS - PRESIÓN ATMOSFÉRICA          hPA  
22611006         PRESIÓN ATMOSFÉRICA          hPa  
22611007         PRESIÓN ATMOSFÉRICA          hPa  

[22611008 rows x 12 columns]

In [34]:
df_presion.count()

CodigoEstacion       22611008
CodigoSensor         22611008
FechaObservacion     22611008
ValorObservado       22611008
NombreEstacion       22611008
Departamento         22611008
Municipio            22611008
ZonaHidrografica     22611008
Latitud              22611008
Longitud             22611008
DescripcionSensor    22611008
UnidadMedida         22611008
dtype: int64

Se observa con la función .count() que el tamaño de valores no nulos es igual al índice del rango del dataframe encontrado en la celda anterior.

In [35]:
df_presion.to_parquet('../CleanDatasets2/presion.parquet')

### Limpieza *DATOS TEMPERATURA*

Los datos de temperatura se distribuyen en 5 archivos csv distintos, todos con las mismas columnas. La ruta de acceso a los archivos es 

- '../Datasets2/Datos_Hidrometeorol_gicos_Crudos_-_Red_de_Estaciones_IDEAM___Temperatura_20241017(*)' Donde * es un valor entre 0 y 4

In [36]:
namesFiles2[2:-1]

['Datos_Hidrometeorol_gicos_Crudos_-_Red_de_Estaciones_IDEAM___Temperatura_20241017(0).csv',
 'Datos_Hidrometeorol_gicos_Crudos_-_Red_de_Estaciones_IDEAM___Temperatura_20241017(1).csv',
 'Datos_Hidrometeorol_gicos_Crudos_-_Red_de_Estaciones_IDEAM___Temperatura_20241017(2).csv',
 'Datos_Hidrometeorol_gicos_Crudos_-_Red_de_Estaciones_IDEAM___Temperatura_20241017(3).csv',
 'Datos_Hidrometeorol_gicos_Crudos_-_Red_de_Estaciones_IDEAM___Temperatura_20241017(4).csv']

In [37]:
df_temp = pd.read_csv(
                        f'../Datasets2/{namesFiles2[2]}',
                        sep=',',
                        nrows=3
)
df_temp

CodigoEstacion  CodigoSensor        FechaObservacion  ValorObservado  \
0        48015050            71  01/04/2024 09:58:00 PM            21.1   
1        15065180            71  01/04/2024 07:46:00 PM            22.3   
2      3502500135            71  01/04/2024 03:56:00 AM            17.0   

              NombreEstacion  Departamento   Municipio     ZonaHidrografica  \
0    AEROPUERTO VASQUEZ COBO      AMAZONAS     LETICIA  AMAZONAS - DIRECTOS   
1  AEROPUERTO ALM. PADILLA -    LA GUAJIRA    RIOHACHA     CARIBE - GUAJIRA   
2  GUAYABETAL POLLO OLIMPICO  CUNDINAMARCA  GUAYABETAL                 META   

     Latitud   Longitud                  DescripcionSensor UnidadMedida  
0  -4.193861 -69.940917  GPRS - TEMPERATURA DEL AIRE A 2 m           °C  
1  11.528444 -72.917722  GPRS - TEMPERATURA DEL AIRE A 2 m           °C  
2   4.225530 -73.814810  GPRS - TEMPERATURA DEL AIRE A 2 m           °C

In [38]:
df_temp = pd.read_csv(
                        f'../Datasets2/{namesFiles2[2]}',
                        sep=',',
                        dtype= {0:'category',1:'category',2:'str',3:'float32',4:'category',5:'category',6:'category',7:'category',8:'category',9:'category',10:'category',11:'category'}         
                    )
df_temp

CodigoEstacion CodigoSensor        FechaObservacion  ValorObservado  \
0           0048015050         0071  01/04/2024 09:58:00 PM       21.100000   
1           0015065180         0071  01/04/2024 07:46:00 PM       22.299999   
2           3502500135         0071  01/04/2024 03:56:00 AM       17.000000   
3           0016015501         0071  01/04/2024 04:30:00 PM       28.500000   
4           0026125710         0071  01/04/2024 06:58:00 AM       20.900000   
...                ...          ...                     ...             ...   
4101303     0015015050         0071  10/16/2024 10:08:00 AM       29.700001   
4101304     0026125710         0071  10/16/2024 05:32:00 AM       18.500000   
4101305     0015015050         0071  10/16/2024 07:20:00 PM       28.000000   
4101306     0017015010         0071  10/16/2024 01:04:00 PM       26.900000   
4101307     0015015050         0071  10/16/2024 02:26:00 PM       31.100000   

                      NombreEstacion  \
0            AEROPUERTO VASQUEZ COBO   
1          AEROPUERTO ALM. PADILLA -   
2          GUAYABETAL POLLO OLIMPICO   
3             AEROPUERTO CAMILO DAZA   
4                AEROPUERTO MATECANA   
...                              ...   
4101303     AEROPUERTO SIMON BOLIVAR   
4101304          AEROPUERTO MATECANA   
4101305     AEROPUERTO SIMON BOLIVAR   
4101306  AEROPUERTO SESQUICENTENARIO   
4101307     AEROPUERTO SIMON BOLIVAR   

                                              Departamento    Municipio  \
0                                                 AMAZONAS      LETICIA   
1                                               LA GUAJIRA     RIOHACHA   
2                                             CUNDINAMARCA   GUAYABETAL   
3                                       NORTE DE SANTANDER       CÚCUTA   
4                                                RISARALDA      PEREIRA   
...                                                    ...          ...   
4101303                                          MAGDALENA  SANTA MARTA   
4101304                                          RISARALDA      PEREIRA   
4101305                                          MAGDALENA  SANTA MARTA   
4101306  ARCHIPIELAGO DE SAN ANDRES PROVIDENCIA Y SANTA...   SAN ANDRÉS   
4101307                                          MAGDALENA  SANTA MARTA   

            ZonaHidrografica      Latitud       Longitud  \
0        AMAZONAS - DIRECTOS  -4.19386111   -69.94091667   
1           CARIBE - GUAJIRA   11.5284444   -72.91772222   
2                       META      4.22553      -73.81481   
3                  CATATUMBO      7.93028      -72.50917   
4                      CAUCA     4.812675   -75.73951944   
...                      ...          ...            ...   
4101303     CARIBE - GUAJIRA   11.1146944  -74.231027778   
4101304                CAUCA     4.812675  -75.739519444   
4101305     CARIBE - GUAJIRA   11.1146944  -74.231027778   
4101306         ISLAS CARIBE    12.587849     -81.701117   
4101307     CARIBE - GUAJIRA   11.1146944  -74.231027778   

                         DescripcionSensor UnidadMedida  
0        GPRS - TEMPERATURA DEL AIRE A 2 m           °C  
1        GPRS - TEMPERATURA DEL AIRE A 2 m           °C  
2        GPRS - TEMPERATURA DEL AIRE A 2 m           °C  
3        GPRS - TEMPERATURA DEL AIRE A 2 m           °C  
4        GPRS - TEMPERATURA DEL AIRE A 2 m           °C  
...                                    ...          ...  
4101303  GPRS - TEMPERATURA DEL AIRE A 2 m           °C  
4101304  GPRS - TEMPERATURA DEL AIRE A 2 m           °C  
4101305  GPRS - TEMPERATURA DEL AIRE A 2 m           °C  
4101306  GPRS - TEMPERATURA DEL AIRE A 2 m           °C  
4101307  GPRS - TEMPERATURA DEL AIRE A 2 m           °C  

[4101308 rows x 12 columns]

In [39]:
df_temp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4101308 entries, 0 to 4101307
Data columns (total 12 columns):
 #   Column             Dtype   
---  ------             -----   
 0   CodigoEstacion     category
 1   CodigoSensor       category
 2   FechaObservacion   object  
 3   ValorObservado     float32 
 4   NombreEstacion     category
 5   Departamento       category
 6   Municipio          category
 7   ZonaHidrografica   category
 8   Latitud            category
 9   Longitud           category
 10  DescripcionSensor  category
 11  UnidadMedida       category
dtypes: category(10), float32(1), object(1)
memory usage: 86.1+ MB


In [40]:
df_temp['FechaObservacion'] = pd.to_datetime(df_temp['FechaObservacion'], format='%m/%d/%Y %H:%M:%S %p')
df_temp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4101308 entries, 0 to 4101307
Data columns (total 12 columns):
 #   Column             Dtype         
---  ------             -----         
 0   CodigoEstacion     category      
 1   CodigoSensor       category      
 2   FechaObservacion   datetime64[ns]
 3   ValorObservado     float32       
 4   NombreEstacion     category      
 5   Departamento       category      
 6   Municipio          category      
 7   ZonaHidrografica   category      
 8   Latitud            category      
 9   Longitud           category      
 10  DescripcionSensor  category      
 11  UnidadMedida       category      
dtypes: category(10), datetime64[ns](1), float32(1)
memory usage: 86.1 MB


In [41]:
df_temp.to_parquet('../CleanDatasets2/Temp_0.parquet')

In [42]:
index_temp = 1
for i in namesFiles2[3:-1]:
    df_temp = pd.read_csv(
                        f'../Datasets2/{i}',
                        sep=',',
                        dtype= {0:'category',1:'category',2:'str',3:'float32',4:'category',5:'category',6:'category',7:'category',8:'category',9:'category',10:'category',11:'category'}         
                    )
    df_temp['FechaObservacion'] = pd.to_datetime(df_temp['FechaObservacion'], format='%m/%d/%Y %H:%M:%S %p')
    df_temp.to_parquet(f'../CleanDatasets2/Temp_{str(index_temp)}.parquet')
    index_temp += 1

Los archivos en formato parquet pueden ser consultados en la carpeta compartida de google drive: https://drive.google.com/drive/folders/1h9_VFIFKbZ8tiavxZM9Nm9koIxJxtCWf?usp=sharing

Se puede acceder a la carpeta compartida desde local instalando la versión de escritorio y buscando la ruta de los archivos.

In [7]:
#namesFiles = os.listdir(r'G:\Mi unidad\Bootcamp\Proyecto_Bootcamp\Data\CleanDatasets')
namesFiles = os.listdir(r'..\CleanDatasets')
print('Primeros 23')
print(namesFiles[0:23])
print('Últimos 23')
print(namesFiles[23:-1])

Primeros 23
['rain_amazonas.parquet', 'rain_antioquia.parquet', 'rain_arauca.parquet', 'rain_atlantico1.parquet', 'rain_atlantico2.parquet', 'rain_bogota1.parquet', 'rain_bogota2.parquet', 'rain_bogota3.parquet', 'rain_bolivar1.parquet', 'rain_bolivar2.parquet', 'rain_boyaca.parquet', 'rain_caldas.parquet', 'rain_caqueta1.parquet', 'rain_caqueta2.parquet', 'rain_casanare.parquet', 'rain_cauca.parquet', 'rain_cesar.parquet', 'rain_choco.parquet', 'rain_choco1.parquet', 'rain_cordoba1.parquet', 'rain_cordoba2.parquet', 'rain_cundinamarca.parquet', 'rain_guainia.parquet']
Últimos 23
['rain_guajira.parquet', 'rain_guaviare.parquet', 'rain_huila.parquet', 'rain_magdalena.parquet', 'rain_meta.parquet', 'rain_narino.parquet', 'rain_narino1.parquet', 'rain_nill.parquet', 'rain_norte_de_santander.parquet', 'rain_putumayo.parquet', 'rain_quindio.parquet', 'rain_risaralda.parquet', 'rain_sanandres1.parquet', 'rain_sanandres2.parquet', 'rain_sanandres3.parquet', 'rain_sanandres4.parquet', 'rain_sa

CleanDatasets contiene los archivos parquet sobre precipitaciones

CleanDatasets2 contiene los archivos parquet sobre temperatura, presión, catálogo de estaciones.

Configuramos rutas para trabajar en local.

In [8]:
path = r'C:\Users\danie\OneDrive\Documents\Bootcamp_Proyecto'                    

In [5]:
path = r'C:\Users\danie\OneDrive\Documents\Bootcamp_Proyecto\CleanDatasets' 

In [3]:
path = r'C:\Users\Josue Florez\Documents\Maria Angelica\Proyectos\Proyecto_bootcamp\Data'

## LIMPIEZA, CARGA Y UNIFICACIÓN DE ARCHIVOS

Lista de archivos:

In [4]:
namesFiles = os.listdir(path)
print(namesFiles)

['CleanDatasets', 'CleanDatasets2']


### Dimensionalidad de los dataset

Estamos interesados en encontrar los valores nulos dentro de los diferentes archivos, en la siguiente celda se calcula un DataFrame que detalla la siguiente información:

- File: Nombre del archivo
- Rows: Cantidad de filas del archivo
- Columns: Cantidad de columnas del archivo
- Size: Tamaño del dataframe, cantidad de celdas
- Nulls: Cantidad de valores nulos

Se observa que ningún archivo contiene valores nulos, además todos tienen la misma cantidad de columnas por tanto es posible formar un solo dataframe

In [5]:
dim = pd.DataFrame(columns=['File','Rows','Columns','Size','Nulls'])

for i in namesFiles:
    path_file = path + "\\" + i
    df_dept2 = pd.read_parquet(
                                path_file
                            )
    row = pd.DataFrame({'File':[i],
                        'Rows':[df_dept2.shape[0]],
                        'Columns':[df_dept2.shape[1]],
                        'Size':[df_dept2.size],
                        'Nulls':[df_dept2.size-df_dept2.count().sum()],
                        'CodigoEstacion': df_dept2['CodigoEstacion'].dtype,
                        'CodigoSensor': df_dept2['CodigoSensor'].dtype,
                        'FechaObservacion': df_dept2['FechaObservacion'].dtype,
                        'ValorObservado': df_dept2['ValorObservado'].dtype,
                        'NombreEstacion': df_dept2['NombreEstacion'].dtype,
                        'Departamento': df_dept2['Departamento'].dtype,
                        'Municipio': df_dept2['Municipio'].dtype,
                        'ZonaHidrografica': df_dept2['ZonaHidrografica'].dtype,
                        'Latitud': df_dept2['Latitud'].dtype,
                        'Longitud': df_dept2['Longitud'].dtype
                        }
                        )
    dim = pd.concat([dim,row], ignore_index=True)

dim

File       Rows Columns        Size Nulls CodigoEstacion  \
0   CleanDatasets  199781152      10  1997811520     0       category   
1  CleanDatasets2   28006224      12   336074688     0       category   

  CodigoSensor FechaObservacion ValorObservado NombreEstacion Departamento  \
0     category   datetime64[ns]        float32       category     category   
1     category   datetime64[ns]        float32       category     category   

  Municipio ZonaHidrografica   Latitud  Longitud  
0  category         category  category  category  
1  category         category  category  category

### Tipos de datos

Estamos interesados en recopilar la información de los diferentes archivos parquet para formar un solo dataframe sobre el cual se pudieran calcular algunas estadísticas exploratorias. Inicialmente, concatenamos los diferentes archivos en un solo dataframe, sin embargo esta operación tardó 80 minutos.

In [6]:
df_part2 = pd.DataFrame()
for i in namesFiles:
    path_file = path + '\\' + i
    df_dept2 = pd.read_parquet(
                                path_file
                            )
    df_part2 = pd.concat([df_part2,df_dept2], ignore_index=True)
    del df_dept2

df_part2

CodigoEstacion CodigoSensor    FechaObservacion  ValorObservado  \
0             0048015050         0240 2019-06-10 03:10:00        0.000000   
1             0044197020         0240 2017-11-12 08:30:00        0.000000   
2             0048015040         0240 2017-04-26 09:10:00        0.000000   
3             0048015040         0240 2019-03-03 06:50:00        0.000000   
4             0048015040         0240 2019-08-03 11:00:00        0.000000   
...                  ...          ...                 ...             ...   
227787371     0048015050         0258 2024-10-16 11:04:00     1000.500000   
227787372     0048015050         0258 2024-10-16 09:08:00     1004.099976   
227787373     0016015501         0258 2024-10-16 07:02:00      972.900024   
227787374     0048015040         0255 2024-10-16 10:40:00      997.400024   
227787375     0011175000         0255 2024-10-16 02:00:00     1006.000000   

                    NombreEstacion        Departamento      Municipio  \
0          AEROPUERTO VASQUEZ COBO            AMAZONAS        LETICIA   
1          VILLAREAL - RIO CAQUETA            AMAZONAS     LA PEDRERA   
2                       PTO NARIÑO            AMAZONAS  PUERTO NARIÑO   
3                       PTO NARIÑO            AMAZONAS  PUERTO NARIÑO   
4                       PTO NARIÑO            AMAZONAS  PUERTO NARIÑO   
...                            ...                 ...            ...   
227787371  AEROPUERTO VASQUEZ COBO            AMAZONAS        LETICIA   
227787372  AEROPUERTO VASQUEZ COBO            AMAZONAS        LETICIA   
227787373   AEROPUERTO CAMILO DAZA  NORTE DE SANTANDER         CÚCUTA   
227787374            PUERTO NARIÑO            AMAZONAS  PUERTO NARIÑO   
227787375                   ATRATO               CHOCO       CÉRTEGUI   

              ZonaHidrografica       Latitud      Longitud  \
0          AMAZONAS - DIRECTOS  -4.193861111  -69.94091667   
1                      CAQUETÁ         -1.31       -69.619   
2          AMAZONAS - DIRECTOS         -3.78       -70.363   
3          AMAZONAS - DIRECTOS         -3.78       -70.363   
4          AMAZONAS - DIRECTOS         -3.78       -70.363   
...                        ...           ...           ...   
227787371  AMAZONAS - DIRECTOS  -4.193861111  -69.94091667   
227787372  AMAZONAS - DIRECTOS  -4.193861111  -69.94091667   
227787373            CATATUMBO       7.93028     -72.50917   
227787374  AMAZONAS - DIRECTOS  -3.780305556  -70.36263889   
227787375      ATRATO - DARIÉN      5.585776    -76.649978   

                    DescripcionSensor UnidadMedida  
0                                 NaN          NaN  
1                                 NaN          NaN  
2                                 NaN          NaN  
3                                 NaN          NaN  
4                                 NaN          NaN  
...                               ...          ...  
227787371  GPRS - PRESIÓN ATMOSFÉRICA          hPA  
227787372  GPRS - PRESIÓN ATMOSFÉRICA          hPA  
227787373  GPRS - PRESIÓN ATMOSFÉRICA          hPA  
227787374         PRESIÓN ATMOSFÉRICA          hPa  
227787375         PRESIÓN ATMOSFÉRICA          hPa  

[227787376 rows x 12 columns]

In [7]:
df_part2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 227787376 entries, 0 to 227787375
Data columns (total 12 columns):
 #   Column             Dtype         
---  ------             -----         
 0   CodigoEstacion     object        
 1   CodigoSensor       object        
 2   FechaObservacion   datetime64[ns]
 3   ValorObservado     float32       
 4   NombreEstacion     object        
 5   Departamento       object        
 6   Municipio          object        
 7   ZonaHidrografica   category      
 8   Latitud            object        
 9   Longitud           object        
 10  DescripcionSensor  category      
 11  UnidadMedida       category      
dtypes: category(3), datetime64[ns](1), float32(1), object(7)
memory usage: 15.1+ GB


Teniendo en cuenta lo anterior y que el tamaño del dataframe en memoria era aprox 14.1GB decidimos construir primero un dataframe basado en las columnas que ofrecen información acerca de la ubicación. Se evidencia una reducción en el tiempo de procesamiento (57s), adicionalmente con el objetivo de conservar la cantidad de registros de cada ubicación, calculamos la columa 'TotalEntries'. Es importante resaltar que la agrupación de todos los archivos indica que existen *199781152* filas.

In [8]:
df_part = pd.DataFrame()
for i in namesFiles:
    path_file = path + '\\' + i
    df_dept_aux = pd.read_parquet(
                                path_file,
                                columns=['CodigoEstacion','NombreEstacion','Departamento','Municipio','ZonaHidrografica','Latitud','Longitud']
                            )
    df_dept = df_dept_aux.groupby(['CodigoEstacion','NombreEstacion','Departamento','Municipio','ZonaHidrografica','Latitud','Longitud'], observed=True).agg({'Longitud':'count'})
    df_dept.columns = ['TotalEntries']
    df_dept = df_dept.reset_index()
    df_dept = df_dept.rename({'CodigoEstacion':'CodSta',
                            'NombreEstacion':'NameSta',
                            'Departamento':'Dept',
                            'Municipio':'City',
                            'ZonaHidrografica':'Zone',
                            'Latitud':'Lat',
                            'Longitud':'Long'
                            }, 
                            axis=1)
    df_dept = df_dept.astype({'CodSta':'str',
                            'NameSta':'str',
                            'Dept':'str',
                            'City':'str',
                            'Zone':'str',
                            'Lat':'float64',
                            'Long':'float64',
                            'TotalEntries':'int64'
                            })
    df_part = pd.concat([df_part,df_dept], ignore_index=True)


df_part.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2335 entries, 0 to 2334
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   CodSta        2335 non-null   object 
 1   NameSta       2335 non-null   object 
 2   Dept          2335 non-null   object 
 3   City          2335 non-null   object 
 4   Zone          2335 non-null   object 
 5   Lat           2335 non-null   float64
 6   Long          2335 non-null   float64
 7   TotalEntries  2335 non-null   int64  
dtypes: float64(2), int64(1), object(5)
memory usage: 146.1+ KB


En la siguiente celda imprimimos la cantidad de valores diferentes por cada columna con el objetivo de identificar incoherencias en los datos y posibles valores duplicados.
Por ejemplo, es posible que un municipio como Bogotá se encuentre escrito de diferentes formas como: 'Bogota', 'Bogotá, D.C.' ...etc pero que corresponden al mismo municipio. Esto puede suceder con los nombres de las estaciones y con demás columnas con formato category

In [9]:
df_part.nunique()

CodSta           949
NameSta         1269
Dept              46
City             521
Zone              34
Lat             1292
Long            1179
TotalEntries    2008
dtype: int64

Note que existen 894 códigos de estación pero existen 1186 nombres de estación, 1191 latitudes y 1098 longitudes. Esto indica que para algunos nombres de estación y coordenadas están escritos de manera diferente para una misma estación. Vamos a hacer un pequeño análisis para identificar estas incongruencias y evitar pérdidas de información.

#### Limpieza de datos correspondiente al nombre y coordenadas de las estaciones

In [10]:
df_part.groupby(['CodSta','NameSta','Lat','Long'], observed=True).agg({'TotalEntries':'sum'}).reset_index()

CodSta                         NameSta       Lat       Long  \
0     0011017020                PR CHOCO: BAGADO  5.412000 -76.418000   
1     0011025501         CARMEN DE ATRATO  - AUT  5.888719 -76.145167   
2     0011027030                        EL SIETE  5.862000 -76.152056   
3     0011027030                  EL SIETE - AUT  5.862000 -76.152056   
4     0011027070                         BORAUDO  5.514611 -76.575694   
...          ...                             ...       ...        ...   
1485  5311500056  UNIVERSIDAD DEL PACIFICO - AUT  3.848308 -76.987017   
1486  5311500121                       LA CUMBRE  3.645194 -76.564750   
1487  5311500121                 LA CUMBRE - AUT  3.645194 -76.564750   
1488  5311500149   COLEGIO VASCO NUÑEZ DE BALBOA  3.884194 -77.049381   
1489  5311700154           PUENTE ZACARIAS - AUT  3.814528 -77.005417   

      TotalEntries  
0            60599  
1           499161  
2             1023  
3           179006  
4               20  
...            ...  
1485        151651  
1486          1341  
1487        176764  
1488         23971  
1489         49591  

[1490 rows x 5 columns]

Se observa que al filtrar por Código de Estación, Nombre de Estación, Latitud y Longitud se encuentran 1375 filas lo cual contrasta con los 894 valores únicos de CodigoEstacion, se evidencia que algunas estaciones tienen el nombre de la estación o las coordenadas escritas de maneras diferentes. Empezaremos por corregir las coordenadas Latitud y Longitud:

##### Coordenadas

Aplicamos una primera aproximación donde redondeamos las coordenadas *Latitud* y *Longitud* a 3 decimales, esto se justifica en que 0.001° equivalen aproximadamente a 100m. Realizamos un conteó de los registros totales que debe coincidir con el valor inicial de *199781152*.

En esta primera celda calculamos el total de registros por cada ubicación, incluimos el nombre de la estación para consultas futuras en el procedimiento de limpieza de las coordenadas.

In [11]:
coordSta_0 = df_part.groupby(['CodSta','NameSta','Lat','Long'], observed=True).agg({'TotalEntries':'sum'}).reset_index()
coordSta_0[['Lat','Long']] = coordSta_0[['Lat','Long']].round(3)
print(f'Total Entries: {coordSta_0['TotalEntries'].agg('sum')}')
coordSta_0

Total Entries: 227787376


CodSta                         NameSta    Lat    Long  TotalEntries
0     0011017020                PR CHOCO: BAGADO  5.412 -76.418         60599
1     0011025501         CARMEN DE ATRATO  - AUT  5.889 -76.145        499161
2     0011027030                        EL SIETE  5.862 -76.152          1023
3     0011027030                  EL SIETE - AUT  5.862 -76.152        179006
4     0011027070                         BORAUDO  5.515 -76.576            20
...          ...                             ...    ...     ...           ...
1485  5311500056  UNIVERSIDAD DEL PACIFICO - AUT  3.848 -76.987        151651
1486  5311500121                       LA CUMBRE  3.645 -76.565          1341
1487  5311500121                 LA CUMBRE - AUT  3.645 -76.565        176764
1488  5311500149   COLEGIO VASCO NUÑEZ DE BALBOA  3.884 -77.049         23971
1489  5311700154           PUENTE ZACARIAS - AUT  3.815 -77.005         49591

[1490 rows x 5 columns]

Al agrupar el dataframe por CodSta, Lat y Long se obtiene un total de 968 filas, de manera que existen Códigos de estación con diferentes coordenadas. Realizaremos ajustes a las coordenadas para obtener una coordenada única para cada estación. 

In [12]:
coordSta_1 = coordSta_0.groupby(['CodSta','Lat','Long'],observed=True).agg({'TotalEntries':'sum'}).reset_index()
print(f'Total Entries: {coordSta_1['TotalEntries'].agg('sum')}')
coordSta_1

Total Entries: 227787376


CodSta    Lat    Long  TotalEntries
0     0011017020  5.412 -76.418         60599
1     0011025501  5.889 -76.145        499161
2     0011027030  5.862 -76.152        180029
3     0011027070  5.515 -76.576        409812
4     0011030010  5.375 -76.613        252329
...          ...    ...     ...           ...
1019  5205500123  1.055 -77.270        102624
1020  5311500056  3.848 -76.987        151651
1021  5311500121  3.645 -76.565        178105
1022  5311500149  3.884 -77.049         23971
1023  5311700154  3.815 -77.005         49591

[1024 rows x 4 columns]

Definimos una función que nos permite encontrar la distancia en Km entre dos ubicaciones en base a las coordenadas.

In [13]:
def Distance(Lat1:float,Long1:float,Lat2:float,Long2:float):
    Lat1, Long1, Lat2, Long2 = map(np.radians, [Lat1, Long1, Lat2, Long2])
    DeltaLong = Long2-Long1
    DeltaLat = Lat2-Lat1
    R = 6371   #Asumiendo un modelo esférico de la tierra  
    return round(2*R*np.arcsin(np.sqrt(np.sin(DeltaLat/2)**2 + ( np.cos(Lat1)*np.cos(Lat2)*np.sin(DeltaLong/2)**2 ))),3)

Calculamos para cada Código de Estación el valor medio de las diferentes coordenadas. Adicionalmente calculamos la columna *Distance* como la distancia de la coordenada registrada al punto coordenado medio.

In [14]:
coordSta_aux = coordSta_1.groupby(['CodSta'], observed=True).agg({'Lat':'mean','Long':'mean'}).round(3)
coordSta_aux = coordSta_aux.reset_index()
coordSta_1 = coordSta_1[['CodSta','Lat','Long','TotalEntries']].merge(coordSta_aux, how='inner', on='CodSta', suffixes=('','_mean'))
coordSta_1['Distance'] = coordSta_1.apply(lambda x: Distance(x['Lat'],x['Long'],x['Lat_mean'],x['Long_mean']), axis=1)
print(f'Total Entries: {coordSta_1['TotalEntries'].agg('sum')}')
coordSta_1

Total Entries: 227787376


CodSta    Lat    Long  TotalEntries  Lat_mean  Long_mean  Distance
0     0011017020  5.412 -76.418         60599     5.412    -76.418     0.000
1     0011025501  5.889 -76.145        499161     5.889    -76.145     0.000
2     0011027030  5.862 -76.152        180029     5.862    -76.152     0.000
3     0011027070  5.515 -76.576        409812     5.515    -76.576     0.000
4     0011030010  5.375 -76.613        252329     5.377    -76.612     0.248
...          ...    ...     ...           ...       ...        ...       ...
1019  5205500123  1.055 -77.270        102624     1.055    -77.270     0.000
1020  5311500056  3.848 -76.987        151651     3.848    -76.987     0.000
1021  5311500121  3.645 -76.565        178105     3.645    -76.565     0.000
1022  5311500149  3.884 -77.049         23971     3.884    -77.049     0.000
1023  5311700154  3.815 -77.005         49591     3.815    -77.005     0.000

[1024 rows x 7 columns]

Consideramos un radio de precisión de las coordenadas de 1.5Km. Los registros de coordenadas con distancias menores a 1.5Km los reemplazamos con el valor medio de las coordenadas. El dataframe *coordSta_2* almacenará las coordenadas de las estaciones que presentan variaciones en sus coordenadas, menores a un radio de 1.5Km desde el punto medio de las coordenadas.

In [15]:
coordSta_2 = coordSta_1[coordSta_1['Distance']<=1.5].copy()
coordSta_2['Lat'] = coordSta_2['Lat_mean']
coordSta_2['Long'] = coordSta_2['Long_mean']
coordSta_2 = coordSta_2.drop(['Lat_mean','Long_mean'], axis=1)
coordSta_2 = coordSta_2.groupby(['CodSta','Lat','Long']).agg({'TotalEntries':'sum'})
coordSta_2 = coordSta_2.reset_index()
print(f'Total Entries: {coordSta_2['TotalEntries'].agg('sum')}')
coordSta_2

Total Entries: 221587143


CodSta    Lat    Long  TotalEntries
0    0011017020  5.412 -76.418         60599
1    0011025501  5.889 -76.145        499161
2    0011027030  5.862 -76.152        180029
3    0011027070  5.515 -76.576        409812
4    0011030010  5.377 -76.612        286859
..          ...    ...     ...           ...
928  5205500123  1.055 -77.270        102624
929  5311500056  3.848 -76.987        151651
930  5311500121  3.645 -76.565        178105
931  5311500149  3.884 -77.049         23971
932  5311700154  3.815 -77.005         49591

[933 rows x 4 columns]

En la siguiente celda listamos los registros que presentan una distancia al punto medio mayor a 1.5Km. Realizamos para cada *CodSta* el cálculo de *%Entries*, esta última columna servirá como criterio para elegir una coordenada única para cada *CodSta*.

In [16]:
coordSta_3_temp = coordSta_1[coordSta_1['Distance']>1.5].copy()

coordSta_aux = coordSta_3_temp.groupby(['CodSta'], observed=True).agg({'TotalEntries':'sum'}).reset_index()
coordSta_3_temp = coordSta_3_temp.merge(coordSta_aux, how='inner', on='CodSta', suffixes=('','_byCodSta'))
coordSta_3_temp['%Entries'] = coordSta_3_temp['TotalEntries']/coordSta_3_temp['TotalEntries_byCodSta']
coordSta_3_temp = coordSta_3_temp.sort_values(['CodSta','%Entries'])
coordSta_3_temp = coordSta_3_temp.drop('TotalEntries_byCodSta', axis=1)
coordSta_3_temp.groupby('CodSta',observed=True).apply(lambda x: x, include_groups=False)

Lat    Long  TotalEntries  Lat_mean  Long_mean  Distance  \
CodSta                                                                       
0011175000 1    5.586 -76.650         73116     5.558    -76.642     3.237   
           0    5.529 -76.634        190936     5.558    -76.642     3.344   
0017015010 2   12.542 -81.731        410011    12.565    -81.716     3.032   
           3   12.588 -81.701        900197    12.565    -81.716     3.032   
0021206780 5    4.619 -74.195         77705     4.618    -74.256     6.762   
           4    4.617 -74.317        316876     4.618    -74.256     6.762   
0021206810 7    4.564 -74.138         71258     4.557    -74.186     5.377   
           6    4.550 -74.233        589417     4.557    -74.186     5.267   
0021208480 8    4.611 -74.178        108128     4.622    -74.164     1.976   
           9    4.633 -74.150        111513     4.622    -74.164     1.976   
0021209920 10   5.106 -73.754          4442     5.149    -73.767     4.993   
           11   5.192 -73.779        887897     5.149    -73.767     4.963   
0024015300 12   5.643 -73.518          1417     5.650    -73.531     1.636   
           13   5.656 -73.544        214887     5.650    -73.531     1.586   
0024035410 14   5.726 -72.921          1409     5.740    -72.916     1.652   
           15   5.753 -72.911        813493     5.740    -72.916     1.548   
0025025240 16   8.543 -74.627         72297     8.544    -74.585     4.620   
           17   8.544 -74.543        127542     8.544    -74.585     4.618   
0054025010 18   4.898 -76.677         64573     4.898    -76.454    24.706   
           19   4.898 -76.232        292585     4.898    -76.454    24.595   
2319500125 21   7.737 -73.049          1380     7.694    -73.114     8.612   
           20   7.650 -73.178        142878     7.694    -73.114     8.584   
2401500086 23   5.554 -73.355          1377     5.498    -73.454    12.603   
           22   5.442 -73.552        191795     5.498    -73.454    12.508   
2403500041 25   5.788 -73.052          1415     5.788    -73.067     1.659   
           24   5.788 -73.082        185970     5.788    -73.067     1.659   
2633500119 26   3.560 -76.457           848     3.722    -76.330    22.872   
           27   3.883 -76.202         97176     3.722    -76.330    22.851   
2633700150 29   4.405 -76.101          1023     4.150    -76.224    31.464   
           28   3.895 -76.347        120951     4.150    -76.224    31.466   
3206500045 30   3.444 -73.744          1335     3.480    -73.742     4.009   
           31   3.516 -73.740        124386     3.480    -73.742     4.009   

               %Entries  
CodSta                   
0011175000 1   0.276900  
           0   0.723100  
0017015010 2   0.312936  
           3   0.687064  
0021206780 5   0.196930  
           4   0.803070  
0021206810 7   0.107856  
           6   0.892144  
0021208480 8   0.492294  
           9   0.507706  
0021209920 10  0.004978  
           11  0.995022  
0024015300 12  0.006551  
           13  0.993449  
0024035410 14  0.001729  
           15  0.998271  
0025025240 16  0.361776  
           17  0.638224  
0054025010 18  0.180797  
           19  0.819203  
2319500125 21  0.009566  
           20  0.990434  
2401500086 23  0.007128  
           22  0.992872  
2403500041 25  0.007551  
           24  0.992449  
2633500119 26  0.008651  
           27  0.991349  
2633700150 29  0.008387  
           28  0.991613  
3206500045 30  0.010619  
           31  0.989381

Teniendo en cuenta las observaciones, elegimos como criterio de selección *%Entries*>0.7, de manera que los registros de coordenadas anómalos serán reemplazados con los registros de coordenadas que cumplan esta condición.

In [17]:
coordSta_aux = coordSta_3_temp[coordSta_3_temp['%Entries']>0.7][['CodSta','Lat','Long']]
coordSta_3 = coordSta_3_temp.merge(coordSta_aux, how='left', on='CodSta', suffixes=('','_sel'))
coordSta_3.loc[coordSta_3['Lat_sel'].notna(),'Lat'] = coordSta_3['Lat_sel']
coordSta_3.loc[coordSta_3['Long_sel'].notna(),'Long'] = coordSta_3['Long_sel']

Los registros que cumplen la condición de tener un registro coordenado con *%Entries*>0.7 se muestran a continuación en el dataframe *coordSta4*:

In [18]:
coordSta_4 = coordSta_3[coordSta_3['Lat_sel'].notna()][coordSta_3.columns[:4]]
coordSta_aux = coordSta_4.groupby(['CodSta','Lat','Long'], observed=True).agg({'TotalEntries':'sum'}).reset_index()
coordSta_4 = coordSta_4.merge(coordSta_aux, how='left', on=['CodSta','Lat','Long'], suffixes=('','_adj'))
coordSta_4 = coordSta_4.drop('TotalEntries', axis=1)
coordSta_4.rename({'TotalEntries_adj':'TotalEntries'}, inplace=True, axis=1)
coordSta_4 = coordSta_4.drop_duplicates().reset_index(drop=True)
coordSta_4

CodSta    Lat    Long  TotalEntries
0   0011175000  5.529 -76.634        264052
1   0021206780  4.617 -74.317        394581
2   0021206810  4.550 -74.233        660675
3   0021209920  5.192 -73.779        892339
4   0024015300  5.656 -73.544        216304
5   0024035410  5.753 -72.911        814902
6   0054025010  4.898 -76.232        357158
7   2319500125  7.650 -73.178        144258
8   2401500086  5.442 -73.552        193172
9   2403500041  5.788 -73.082        187385
10  2633500119  3.883 -76.202         98024
11  2633700150  3.895 -76.347        121974
12  3206500045  3.516 -73.740        125721

Por otra parte, a continuación se lista los registros que no cumplen dicha condición. Para estas estaciones realizamos una busqueda manual en GoogleEarth para identificar las coordenadas correctas de las estaciones.

In [19]:
coordSta_5 = coordSta_3[coordSta_3['Lat_sel'].isna()][coordSta_3.columns[:4]]
coordSta_aux = coordSta_5.merge(coordSta_0, how='left', on=['CodSta','Lat','Long','TotalEntries'])
coordSta_aux

CodSta     Lat    Long  TotalEntries                        NameSta
0  0017015010  12.542 -81.731        410011  APTO SESQUICENTENARIO TX GPRS
1  0017015010  12.588 -81.701        900197    AEROPUERTO SESQUICENTENARIO
2  0021208480   4.611 -74.178        108128                        KENNEDY
3  0021208480   4.633 -74.150        111513                KENNEDY - FOPAE
4  0025025240   8.543 -74.627         72297                            NaN
5  0025025240   8.544 -74.543        127542                       MAJAGUAL

Las coordenadas correctas de cada sitio son: 

    AEROPUERTO SESQUICENTENARIO  '0017015010':[12.588,-81.701]
    KENNEDY                      '0021208480':[4.633,-74.150]
    MAJAGUAL                     '0025025240':[8.543,-74.627]

In [20]:
coordSta_5 = coordSta_3[coordSta_3['Lat_sel'].isna()][coordSta_3.columns[:4]]
coordSta_aux = pd.DataFrame({'0017015010':[12.588,-81.701],
                             '0021208480':[4.633,-74.150],
                             '0025025240':[8.543,-74.627]
                             }, index=coordSta_5.columns[1:-1])
coordSta_aux = coordSta_aux.transpose().reset_index(names='CodSta')
coordSta_5 = coordSta_5.merge(coordSta_aux, how='left', on='CodSta', suffixes=('','_adj'))
coordSta_5['Lat'] = coordSta_5['Lat_adj']
coordSta_5['Long'] = coordSta_5['Long_adj']
coordSta_5 = coordSta_5.groupby(['CodSta','Lat','Long'],observed=True).agg({'TotalEntries':'sum'}).reset_index()
coordSta_5

CodSta     Lat    Long  TotalEntries
0  0017015010  12.588 -81.701       1310208
1  0021208480   4.633 -74.150        219641
2  0025025240   8.543 -74.627        199839

Finalmente, concatenamos los DataFrames que contienen las correcciones aplicadas en cada criterio usado:

    coordSta_2 : Radio de aproximación < 1.5Km
    coordSta_4 : Radio de aproximación > 1.5Km y registros con frecuencias relativas mayores a 0.7
    coordSta_5 : Corrección manual

Adicionalmente, realizamos una sumatoria del total de registros para cada CodSta con el fin de evaluar la integridad de la operación de limpieza. El valor inicial de registros es de *199781152*.

In [21]:
coordSta_f = pd.concat([coordSta_2,coordSta_4,coordSta_5], ignore_index=True)
print(f'Total Entries: {coordSta_f['TotalEntries'].agg('sum')}')
coordSta_f

Total Entries: 227787376


CodSta     Lat    Long  TotalEntries
0    0011017020   5.412 -76.418         60599
1    0011025501   5.889 -76.145        499161
2    0011027030   5.862 -76.152        180029
3    0011027070   5.515 -76.576        409812
4    0011030010   5.377 -76.612        286859
..          ...     ...     ...           ...
944  2633700150   3.895 -76.347        121974
945  3206500045   3.516 -73.740        125721
946  0017015010  12.588 -81.701       1310208
947  0021208480   4.633 -74.150        219641
948  0025025240   8.543 -74.627        199839

[949 rows x 4 columns]

##### Nombre Estaciones

Ahora nos interesa corregir los nombres de las estaciones, ya que se evidencia que algunos Códigos de Estación tienen asociados diferentes nombres.

In [22]:
NameSta_0 = coordSta_0.groupby(['CodSta','NameSta'],observed=True).agg({'TotalEntries':'sum'}).reset_index()
NameSta_0['SizeString'] = NameSta_0['NameSta'].apply(lambda x: len(x))
NameSta_0

CodSta                         NameSta  TotalEntries  SizeString
0     0011017020                PR CHOCO: BAGADO         60599          16
1     0011025501         CARMEN DE ATRATO  - AUT        499161          23
2     0011027030                        EL SIETE          1023           8
3     0011027030                  EL SIETE - AUT        179006          14
4     0011027070                         BORAUDO        409812           7
...          ...                             ...           ...         ...
1301  5311500056  UNIVERSIDAD DEL PACIFICO - AUT        151651          30
1302  5311500121                       LA CUMBRE          1341           9
1303  5311500121                 LA CUMBRE - AUT        176764          15
1304  5311500149   COLEGIO VASCO NUÑEZ DE BALBOA         23971          29
1305  5311700154           PUENTE ZACARIAS - AUT         49591          21

[1306 rows x 4 columns]

Filtramos inicialmente los registros con un único Nombre de estación para cada código de estación. 

In [23]:
NameSta_1 = NameSta_0[~NameSta_0[['CodSta']].duplicated(keep=False)][NameSta_0.columns[:-1]].reset_index(drop=True)
NameSta_1

CodSta                              NameSta  TotalEntries
0    0011017020                     PR CHOCO: BAGADO         60599
1    0011025501              CARMEN DE ATRATO  - AUT        499161
2    0011027070                              BORAUDO        409812
3    0011037030  LA LOMA PUEBLO NUEVO - En Siniestro        312861
4    0011050020                       PR CHOCO: BETE        152099
..          ...                                  ...           ...
590  5204700100             PUENTE LOS DOLORES - AUT         98176
591  5205500123                  LAS IGLESIAS  - AUT        102624
592  5311500056       UNIVERSIDAD DEL PACIFICO - AUT        151651
593  5311500149        COLEGIO VASCO NUÑEZ DE BALBOA         23971
594  5311700154                PUENTE ZACARIAS - AUT         49591

[595 rows x 3 columns]

Dado que el nombre de la estación no es una variable de interés reelevante para el proyecto, elegimos como criterio para seleccionar un nombre de estación tener una probabilidad de ocurrencia mayor a 0.5

In [24]:
NameSta_2 = NameSta_0[NameSta_0[['CodSta']].duplicated(keep=False)].reset_index(drop=True)
NameSta_2['%Entries'] = NameSta_2.groupby('CodSta', observed=True)[['TotalEntries']].transform(lambda x: x/x.sum())
NameSta_2aux = NameSta_2.loc[NameSta_2['%Entries']>0.5,['CodSta','NameSta']]
NameSta_2 = NameSta_2.merge(NameSta_2aux, how='left', on='CodSta', suffixes=('','_adj'))
NameSta_2.loc[NameSta_2['NameSta_adj'].notna(),'NameSta'] = NameSta_2['NameSta_adj']
NameSta_3 = NameSta_2[NameSta_2['NameSta_adj'].notna()][NameSta_2.columns[:3]]
NameSta_3 = NameSta_3.groupby(['CodSta','NameSta'],observed=True).agg({'TotalEntries':'sum'}).reset_index()
NameSta_3

CodSta                       NameSta  TotalEntries
0    0011027030                EL SIETE - AUT        180029
1    0011030010            PR CHOCO: CERTEGUI        286859
2    0011035030  PR CHOCO: UNION PANAMERICANA        129066
3    0011045010      PR CHOCO: APTO EL CARANO        297027
4    0011047040                  QUIBDO - AUT        595199
..          ...                           ...           ...
348  3505500061                  MEDINA - AUT        142994
349  3519700092                    MANI - AUT        131678
350  3602700101            SAN SALVADOR - AUT        157685
351  3706500109         ARAUCA TERMINAL - AUT        163363
352  5311500121               LA CUMBRE - AUT        178105

[353 rows x 3 columns]

Adicionalmente, realizamos una sumatoria del total de registros para cada CodSta con el fin de evaluar la integridad de la operación de limpieza. El valor inicial de registros es de *199781152*.

In [25]:
NameSta_f = pd.concat([NameSta_1,NameSta_3], ignore_index=True)
print(f'Total Entries: {NameSta_f['TotalEntries'].agg('sum')}')
NameSta_f

Total Entries: 226865713


CodSta                              NameSta  TotalEntries
0    0011017020                     PR CHOCO: BAGADO         60599
1    0011025501              CARMEN DE ATRATO  - AUT        499161
2    0011027070                              BORAUDO        409812
3    0011037030  LA LOMA PUEBLO NUEVO - En Siniestro        312861
4    0011050020                       PR CHOCO: BETE        152099
..          ...                                  ...           ...
943  3505500061                         MEDINA - AUT        142994
944  3519700092                           MANI - AUT        131678
945  3602700101                   SAN SALVADOR - AUT        157685
946  3706500109                ARAUCA TERMINAL - AUT        163363
947  5311500121                      LA CUMBRE - AUT        178105

[948 rows x 3 columns]

Recopilamos los resultados de la limpieza de las coordenadas y los nombres de las estaciones en un solo dataframe.

In [26]:
df_part_f = NameSta_f.merge(coordSta_f, how='inner', on='CodSta', suffixes=('','_coord'))
print(f'Total entries of NameSta: {df_part_f['TotalEntries'].sum()}')
print(f'Total entries of CoordSta: {df_part_f['TotalEntries_coord'].sum()}')
df_part_f = df_part_f.drop(['TotalEntries','TotalEntries_coord'],axis=1)
df_part_f

Total entries of NameSta: 226865713
Total entries of CoordSta: 226865713


CodSta                              NameSta    Lat    Long
0    0011017020                     PR CHOCO: BAGADO  5.412 -76.418
1    0011025501              CARMEN DE ATRATO  - AUT  5.889 -76.145
2    0011027070                              BORAUDO  5.515 -76.576
3    0011037030  LA LOMA PUEBLO NUEVO - En Siniestro  5.585 -76.753
4    0011050020                       PR CHOCO: BETE  5.995 -76.780
..          ...                                  ...    ...     ...
943  3505500061                         MEDINA - AUT  4.508 -73.348
944  3519700092                           MANI - AUT  4.816 -72.309
945  3602700101                   SAN SALVADOR - AUT  6.228 -71.618
946  3706500109                ARAUCA TERMINAL - AUT  7.064 -70.733
947  5311500121                      LA CUMBRE - AUT  3.645 -76.565

[948 rows x 4 columns]

#### Limpieza de las columnas Municipio, Departamento y ZonaHidrografica

Realizamos un conteo de los valores únicos de cada columna.

In [27]:
Ubic_0 = df_part[['CodSta','Dept','City','Zone']].copy()
Ubic_0 = Ubic_0.drop_duplicates().reset_index(drop=True)
Ubic_0.nunique()

CodSta    949
Dept       46
City      521
Zone       34
dtype: int64

##### Departamentos

Realizamos una limpieza sobre los nombres de los departamentos, esto lo realizamos de manera manual debido a que son correcciones ortográficas.

In [28]:
Dept = Ubic_0['Dept'].drop_duplicates().reset_index(drop=True)
Dept = Dept.astype('str')
print(list(Dept))

['AMAZONAS', 'ANTIOQUIA', 'ARAUCA', 'ATLÁNTICO', 'ATLANTICO', 'BOGOTÁ', 'BOGOTA D.C.', 'BOGOTA', 'CUNDINAMARCA', 'BOLIVAR', 'BOLÍVAR', 'SUCRE', 'BOYACÁ', 'CALDAS', 'CAQUETA', 'CAQUETÁ', 'CASANARE', 'CAUCA', 'CESAR', 'CHOCO', 'CHOCÓ', 'CÓRDOBA', 'CORDOBA', 'GUAINÍA', 'LA GUAJIRA', 'GUAVIARE', 'HUILA', 'MAGDALENA', 'META', 'NARINO', 'NARIÑO', '<nil>', 'NORTE DE SANTANDER', 'PUTUMAYO', 'QUINDÍO', 'RISARALDA', 'SAN ANDRÉS PROVIDENCIA', 'ARCHIPIÉLAGO DE SAN ANDRES PROVIDENCIA Y SANTA CATALINA', 'ARCHIPIELAGO DE SAN ANDRES PROVIDENCIA Y SANTA CATALINA', 'ARCHIPIELAGO DE SAN ANDRES, PROVIDENCIA Y SANTA CATALINA', 'SANTANDER', 'TOLIMA', 'VALLE DEL CAUCA', 'VAUPES', 'VAUPÉS', 'VICHADA']


In [29]:
Ubic_0.replace({'Dept':r'ATL.NTICO'},{'Dept':'ATLÁNTICO'},regex=True, inplace=True)
Ubic_0.replace({'Dept':r'.*BOG.*'},{'Dept':'BOGOTÁ, D.C.'},regex=True, inplace=True)
Ubic_0.replace({'Dept':r'BOL.VAR'},{'Dept':'BOLÍVAR'},regex=True, inplace=True)
Ubic_0.replace({'Dept':r'CAQUET.*'},{'Dept':'CAQUETÁ'},regex=True, inplace=True)
Ubic_0.replace({'Dept':r'CHOC.'},{'Dept':'CHOCÓ'},regex=True, inplace=True)
Ubic_0.replace({'Dept':r'C.RDOBA'},{'Dept':'CÓRDOBA'},regex=True, inplace=True)
Ubic_0.replace({'Dept':r'.*SAN AND.*'},{'Dept':'ARCHIPIÉLAGO DE SAN ANDRÉS, PROVIDENCIA Y SANTA CATALINA'},regex=True, inplace=True)
Ubic_0.replace({'Dept':'NARINO'},{'Dept':'NARIÑO'},regex=True, inplace=True)
Ubic_0.replace({'Dept':'VAUPES'},{'Dept':'VAUPÉS'},regex=True, inplace=True)
Dept = Ubic_0['Dept'].drop_duplicates().reset_index(drop=True)
Dept = Dept.astype('str')
print(list(Dept))


['AMAZONAS', 'ANTIOQUIA', 'ARAUCA', 'ATLÁNTICO', 'BOGOTÁ, D.C.', 'CUNDINAMARCA', 'BOLÍVAR', 'SUCRE', 'BOYACÁ', 'CALDAS', 'CAQUETÁ', 'CASANARE', 'CAUCA', 'CESAR', 'CHOCÓ', 'CÓRDOBA', 'GUAINÍA', 'LA GUAJIRA', 'GUAVIARE', 'HUILA', 'MAGDALENA', 'META', 'NARIÑO', '<nil>', 'NORTE DE SANTANDER', 'PUTUMAYO', 'QUINDÍO', 'RISARALDA', 'ARCHIPIÉLAGO DE SAN ANDRÉS, PROVIDENCIA Y SANTA CATALINA', 'SANTANDER', 'TOLIMA', 'VALLE DEL CAUCA', 'VAUPÉS', 'VICHADA']


Vemos en la siguiente celda que las columnas Dept, City y Zone contienen un valor nulo nil por tanto, vamos a obtener la información de las coordenadas asociadas a estos códigos de estación para recuperar los valores correctos de las columnas con estos datos nulos.

In [30]:
Ubic_D1 = Ubic_0[Ubic_0['Dept']=='<nil>']
Ubic_D1.merge(df_part_f, how='inner', on='CodSta')

CodSta   Dept   City       Zone                               NameSta  \
0  0016027120  <nil>  <nil>  CATATUMBO                SAN JAVIER - RIO ZULIA   
1  0023157050  <nil>  <nil>      <nil>                       BARRANCABERMEJA   
2  0035217080  <nil>  <nil>      <nil>  PERENCO: LA CABANA - TERMO ELECTRICA   
3  0035227020  <nil>  <nil>      <nil>               PERENCO: OROCUE PIPESCA   
4  0035237040  <nil>  <nil>      <nil>               PERENCO: TRINIDAD METEO   
5  0088112901  <nil>  <nil>      <nil>    ECI JULIO GARAVITO EST. EN PRUEBAS   

     Lat    Long  
0  7.834 -72.650  
1  7.060 -73.877  
2  5.438 -72.455  
3  4.783 -71.345  
4  5.419 -71.666  
5  0.000   0.000

Realizando una busquedo en google earth encontramos la siguiente información:
    {0016027120: {'Dept':'NORTE DE SANTANDER','City':'DURANIA', 'Zone':'CATATUMBO}}
    {0023157050: {'Dept':'SANTANDER','City':'BARRANCABERMEJA', 'Zone':'MAGDALENA MEDIO'}}
    {0035217080: {'Dept':'CASANARE','City':'YOPAL', 'Zone':'META'}}
    {0035227020: {'Dept':'CASANARE','City':'OROCUÉ', 'Zone':'META'}}
    {0035237040: {'Dept':'CASANARE','City':'TRINIDAD', 'Zone':'META'}}
    {0088112901: {'Dept':'BOGOTÁ, D.C.','City':'BOGOTÁ, D.C.', 'Zone':'MAGDALENA MEDIO'}}

In [31]:
Ubic_D2 =pd.DataFrame([{'CodSta':'0016027120','Dept':'NORTE DE SANTANDER','City':'DURANIA', 'Zone':'CATATUMBO'},
    {'CodSta':'0023157050','Dept':'SANTANDER','City':'BARRANCABERMEJA', 'Zone':'MEDIO MAGDALENA'},
    {'CodSta':'0035217080','Dept':'CASANARE','City':'YOPAL', 'Zone':'META'},
    {'CodSta':'0035227020','Dept':'CASANARE','City':'OROCUÉ', 'Zone':'META'},
    {'CodSta':'0035237040','Dept':'CASANARE, D.C.','City':'TRINIDAD', 'Zone':'META'},
    {'CodSta':'0088112901','Dept':'BOGOTÁ, D.C.','City':'BOGOTÁ, D.C.', 'Zone':'MEDIO MAGDALENA'},    
    ]
)
Ubic_D2

CodSta                Dept             City             Zone
0  0016027120  NORTE DE SANTANDER          DURANIA        CATATUMBO
1  0023157050           SANTANDER  BARRANCABERMEJA  MEDIO MAGDALENA
2  0035217080            CASANARE            YOPAL             META
3  0035227020            CASANARE           OROCUÉ             META
4  0035237040      CASANARE, D.C.         TRINIDAD             META
5  0088112901        BOGOTÁ, D.C.     BOGOTÁ, D.C.  MEDIO MAGDALENA

In [32]:
Ubic_1 = Ubic_0.merge(Ubic_D2, how='left', on='CodSta', suffixes=('','_nil'))
Ubic_1.loc[Ubic_1['Dept_nil'].notna(),'Dept'] = Ubic_1['Dept_nil']
Ubic_1.loc[Ubic_1['City_nil'].notna(),'City'] = Ubic_1['City_nil']
Ubic_1.loc[Ubic_1['Zone_nil'].notna(),'Zone'] = Ubic_1['Zone_nil']
Ubic_1 = Ubic_1[Ubic_1.columns[:-3]]
Ubic_1

CodSta       Dept                  City                 Zone
0     0044197020   AMAZONAS            LA PEDRERA              CAQUETÁ
1     0048015010   AMAZONAS               LETICIA                <nil>
2     0048015040   AMAZONAS         PUERTO NARIÑO  AMAZONAS - DIRECTOS
3     0048015050   AMAZONAS               LETICIA  AMAZONAS - DIRECTOS
4     0011057020  ANTIOQUIA      VIGIA DEL FUERTE      ATRATO - DARIÉN
...          ...        ...                   ...                  ...
1099  0024035380     BOYACÁ                GUICÁN             SOGAMOSO
1100  0027015300  ANTIOQUIA              MEDELLÍN                NECHÍ
1101  0029045150    CÓRDOBA             TIERRALTA                 SINÚ
1102  0047035030   PUTUMAYO      PUERTO LEGUÍZAMO              CAQUETÁ
1103  0056019010      CHOCÓ  BAHÍA SOLANO (MUTIS)  PACÍFICO - DIRECTOS

[1104 rows x 4 columns]

In [33]:
print(Ubic_1['Dept'].unique())

['AMAZONAS' 'ANTIOQUIA' 'ARAUCA' 'ATLÁNTICO' 'BOGOTÁ, D.C.' 'CUNDINAMARCA'
 'BOLÍVAR' 'SUCRE' 'BOYACÁ' 'CALDAS' 'CAQUETÁ' 'CASANARE' 'CAUCA' 'CESAR'
 'CHOCÓ' 'CÓRDOBA' 'GUAINÍA' 'LA GUAJIRA' 'GUAVIARE' 'HUILA' 'MAGDALENA'
 'META' 'NARIÑO' 'NORTE DE SANTANDER' 'SANTANDER' 'CASANARE, D.C.'
 'PUTUMAYO' 'QUINDÍO' 'RISARALDA'
 'ARCHIPIÉLAGO DE SAN ANDRÉS, PROVIDENCIA Y SANTA CATALINA' 'TOLIMA'
 'VALLE DEL CAUCA' 'VAUPÉS' 'VICHADA']


##### Municipios

Nos interesamos inicialmente en encontrar aquellas estaciones que tienen asociados más de un municipio.

In [34]:
Ubic_M1 = Ubic_1[['CodSta','City']].drop_duplicates()
Ubic_M1 = Ubic_M1[Ubic_M1['CodSta'].duplicated(keep=False)].reset_index(drop=True)
Ubic_M1 = Ubic_M1.merge(coordSta_f[['CodSta','Lat','Long']], how='inner', on='CodSta', suffixes=('','_adj'))
Ubic_M1 = Ubic_M1.sort_values(['CodSta','City']).reset_index(drop=True)
Ubic_M1 = Ubic_M1.groupby(['CodSta','Lat','Long'], observed=True)['City'].agg(list).reset_index()
Ubic_M1['Count'] = Ubic_M1['City'].apply(lambda x: len(x))
print(f'Maximum number of different cities by CodSta: {Ubic_M1['Count'].max()}')
Ubic_M1
#184

Maximum number of different cities by CodSta: 3


CodSta     Lat    Long  \
0   0011175000   5.529 -76.634   
1   0017025020  13.360 -81.358   
2   0021201200   4.343 -74.184   
3   0021201580   4.446 -74.155   
4   0021201980   4.579 -74.053   
..         ...     ...     ...   
78  2120700038   4.562 -74.149   
79  2120700154   4.553 -74.108   
80  2401500086   5.442 -73.552   
81  2633500119   3.883 -76.202   
82  2633700150   3.895 -76.347   

                                                 City  Count  
0                               [CÉRTEGUI, EL ATRATO]      2  
1   [SAN ANDRES Y  PROVIDENCIA, SAN ANDRES Y  PROV...      2  
2               [BOGOTA D.C, BOGOTA, D.C, BOGOTÁ D.C]      3  
3               [BOGOTA D.C, BOGOTA, D.C, BOGOTÁ D.C]      3  
4               [BOGOTA D.C, BOGOTA, D.C, BOGOTÁ D.C]      3  
..                                                ...    ...  
78              [BOGOTA D.C, BOGOTA, D.C, BOGOTÁ D.C]      3  
79                           [BOGOTA D.C, BOGOTÁ D.C]      2  
80                                    [SAMACÁ, TUNJA]      2  
81                                    [BUGA, PALMIRA]      2  
82                                     [BUGA, ZARZAL]      2  

[83 rows x 5 columns]

In [35]:
Ubic_M1['City'].drop_duplicates().reset_index(drop=True)

0                                 [CÉRTEGUI, EL ATRATO]
1     [SAN ANDRES Y  PROVIDENCIA, SAN ANDRES Y  PROV...
2                 [BOGOTA D.C, BOGOTA, D.C, BOGOTÁ D.C]
3                              [BOGOTA D.C, BOGOTÁ D.C]
4                             [BOGOTA, D.C, BOGOTÁ D.C]
5                           [ARMERO, ARMERO (GUAYABAL)]
6                                 [CHAPARRAL, PLANADAS]
7                                      [ACHÍ, MAJAGUAL]
8                               [ANSERMANUEVO, ARGELIA]
9                                    [TÁMESIS, VENECIA]
10                      [MALLAMA, MALLAMA (PIEDRANCHA)]
11                            [PATÍA, PATÍA (EL BORDO)]
12                        [NÓVITA, SAN JOSÉ DEL PALMAR]
13                                   [RIO IRO, RIO IRÓ]
14               [ALTO BAUDÓ, ALTO BAUDÓ (PIE DE PATO)]
15                 [BAHÍA SOLANO, BAHÍA SOLANO (MUTIS)]
16                            [BOGOTA D.C, BOGOTA, D.C]
17                                      [SAMACÁ,

Se podría realizar un reemplazo inicial en base a las siguientes condiciones:

    - Eliminar caracteres dentro de paréntesis
    - Hacer correcciones ortográficas sobre: 'San Andrés, providencia y santa catalina', 'Bogotá, D.C.', y 'Rio Iró' 

Los cambios los aplicaremos sobre el dataframe Ubic_1

In [36]:
Ubic_1.replace({'City':r'.\(.*\)'},{'City':''},regex=True, inplace=True)
Ubic_1.replace({'City':r'.*BOG.*'},{'City':'BOGOTÁ, D.C.'},regex=True, inplace=True)
Ubic_1.replace({'City':r'.*SAN AND.*'},{'City':'ARCHIPIÉLAGO DE SAN ANDRÉS, PROVIDENCIA Y SANTA CATALINA'},regex=True, inplace=True)
Ubic_1.replace({'City':r'RIO IR.*'},{'City':'RIO IRÓ'},regex=True, inplace=True)

De nuevo aplicamos el filtrado anterior para listar las estaciones que presentan mas de un municipio en los registros. Se evidencian 9 estaciones con este tipo de diferencias.

In [37]:
Ubic_M2 = Ubic_1[['CodSta','City']].drop_duplicates()
Ubic_M2 = Ubic_M2[Ubic_M2['CodSta'].duplicated(keep=False)].reset_index(drop=True)
Ubic_M2 = Ubic_M2.merge(coordSta_f[['CodSta','Lat','Long']], how='inner', on='CodSta', suffixes=('','_adj'))
Ubic_M2 = Ubic_M2.sort_values(['CodSta','City']).reset_index(drop=True)
Ubic_M2 = Ubic_M2.groupby(['CodSta','Lat','Long'], observed=True)['City'].agg(list).reset_index()
Ubic_M2['Count'] = Ubic_M2['City'].apply(lambda x: len(x))
print(f'Maximum number of different cities by CodSta: {Ubic_M2['Count'].max()}')
Ubic_M2

Maximum number of different cities by CodSta: 2


CodSta    Lat    Long                           City  Count
0  0011175000  5.529 -76.634          [CÉRTEGUI, EL ATRATO]      2
1  0022045502  3.950 -75.667          [CHAPARRAL, PLANADAS]      2
2  0025025240  8.543 -74.627               [ACHÍ, MAJAGUAL]      2
3  0026115090  4.776 -76.144        [ANSERMANUEVO, ARGELIA]      2
4  0026205501  5.626 -75.705             [TÁMESIS, VENECIA]      2
5  0054025010  4.898 -76.232  [NÓVITA, SAN JOSÉ DEL PALMAR]      2
6  2401500086  5.442 -73.552                [SAMACÁ, TUNJA]      2
7  2633500119  3.883 -76.202                [BUGA, PALMIRA]      2
8  2633700150  3.895 -76.347                 [BUGA, ZARZAL]      2

Para resolver este conflicto, realizamos una busqueda en Google Earth para encontrar el municipio relacionado con las coordenadas de la estación. Las correcciones son las siguientes:

    - 0011175000 EL ATRATO
    - 0022045502 CHAPARRAL
    - 0025025240 MAJAGUAL
    - 0026115090 ANSERMANUEVO
    - 0026205501 TÁMESIS
    - 0054025010 SAN JOSÉ DEL PALMAR
    - 2401500086 SAMACÁ
    - 2633500119 BUGA
    - 2633700150 BUGA

In [38]:
Ubic_1.loc[Ubic_1['CodSta']=='0011175000','City'] = 'EL ATRATO'
Ubic_1.loc[Ubic_1['CodSta']=='0022045502','City'] = 'CHAPARRAL'
Ubic_1.loc[Ubic_1['CodSta']=='0025025240','City'] = 'MAJAGUAL'
Ubic_1.loc[Ubic_1['CodSta']=='0026115090','City'] = 'ANSERMANUEVO'
Ubic_1.loc[Ubic_1['CodSta']=='0026205501','City'] = 'TÁMESIS'
Ubic_1.loc[Ubic_1['CodSta']=='0054025010','City'] = 'SAN JOSÉ DEL PALMAR'
Ubic_1.loc[Ubic_1['CodSta']=='2401500086','City'] = 'SAMACÁ'
Ubic_1.loc[Ubic_1['CodSta']=='2633500119','City'] = 'BUGA'
Ubic_1.loc[Ubic_1['CodSta']=='2633700150','City'] = 'BUGA'
Ubic_1[['CodSta','City']].drop_duplicates().reset_index(drop=True)

CodSta              City
0    0044197020        LA PEDRERA
1    0048015010           LETICIA
2    0048015040     PUERTO NARIÑO
3    0048015050           LETICIA
4    0011057020  VIGIA DEL FUERTE
..          ...               ...
944  0024035380            GUICÁN
945  0027015300          MEDELLÍN
946  0029045150         TIERRALTA
947  0047035030  PUERTO LEGUÍZAMO
948  0056019010      BAHÍA SOLANO

[949 rows x 2 columns]

Con las modificaciones realizadas se encuentra que cada estación ahora queda asociada a un municipio.

##### Zonas hidrográficas

En la siguiente celda listamos las zonas hidrográficas. Sin embargo, dichas delimitaciones no representan adecuadamente las zonas hidrográficas nacionales, ya que en los registros no parece existir un criterio claro para delimitar dichas zonas. Por esta razón vamos a excluir esta columna del análisis.

In [39]:
Ubic_1['Zone'].drop_duplicates().reset_index(drop=True).sort_values().values[:]

array(['<nil>', 'ALTO MAGDALENA', 'AMAZONAS - DIRECTOS', 'ARAUCA',
       'ATRATO - DARIÉN', 'BAJO MAGDALENA',
       'BAJO MAGDALENA- CAUCA -SAN JORGE', 'BAUDÓ - DIRECTOS PACIFICO',
       'CAGUÁN', 'CAQUETÁ', 'CARIBE - GUAJIRA', 'CARIBE - LITORAL',
       'CASANARE', 'CATATUMBO', 'CAUCA', 'CESAR', 'GUAVIARE', 'INÍRIDA',
       'ISLAS CARIBE', 'MEDIO MAGDALENA', 'META', 'MIRA', 'NECHÍ',
       'ORINOCO', 'PACÍFICO - DIRECTOS', 'PATÍA', 'PUTUMAYO', 'SALDAÑA',
       'SAN JUÁN', 'SINÚ', 'SOGAMOSO', 'TAPAJE - DAGUA - DIRECTOS',
       'VAUPES', 'VICHADA'], dtype=object)

In [40]:
Ubic_2 = Ubic_1.drop('Zone', axis=1).drop_duplicates().reset_index(drop=True)
Ubic_2


CodSta       Dept              City
0    0044197020   AMAZONAS        LA PEDRERA
1    0048015010   AMAZONAS           LETICIA
2    0048015040   AMAZONAS     PUERTO NARIÑO
3    0048015050   AMAZONAS           LETICIA
4    0011057020  ANTIOQUIA  VIGIA DEL FUERTE
..          ...        ...               ...
946  0024035380     BOYACÁ            GUICÁN
947  0027015300  ANTIOQUIA          MEDELLÍN
948  0029045150    CÓRDOBA         TIERRALTA
949  0047035030   PUTUMAYO  PUERTO LEGUÍZAMO
950  0056019010      CHOCÓ      BAHÍA SOLANO

[951 rows x 3 columns]

##### Estaciones registradas con diferentes departamentos

A continuación listamos un pequeño grupo de registros que presentan incoherencias en la columna Dept. Para corregir las incoherencias vamos a realizar una busqueda en internet para determinar el departamento correcto de cada municipio.

In [41]:
Ubic_2[Ubic_2['CodSta'].duplicated(keep=False)].sort_values('CodSta').reset_index(drop=True)

CodSta          Dept          City
0  0025025240       BOLÍVAR      MAJAGUAL
1  0025025240         SUCRE      MAJAGUAL
2  2120500204  BOGOTÁ, D.C.  BOGOTÁ, D.C.
3  2120500204  CUNDINAMARCA  BOGOTÁ, D.C.

Los departamentos correctos de las estaciones son:

    - 0025025240 : SUCRE
    - 2120500204 : BOGOTÁ, D.C.

In [42]:
Ubic_2.loc[Ubic_2['CodSta']=='0025025240','Dept'] = 'SUCRE'
Ubic_2.loc[Ubic_2['CodSta']=='2120500204','Dept'] = 'BOGOTÁ, D.C.'
Ubic_f = Ubic_2.drop_duplicates().reset_index(drop=True)
Ubic_f

CodSta       Dept              City
0    0044197020   AMAZONAS        LA PEDRERA
1    0048015010   AMAZONAS           LETICIA
2    0048015040   AMAZONAS     PUERTO NARIÑO
3    0048015050   AMAZONAS           LETICIA
4    0011057020  ANTIOQUIA  VIGIA DEL FUERTE
..          ...        ...               ...
944  0024035380     BOYACÁ            GUICÁN
945  0027015300  ANTIOQUIA          MEDELLÍN
946  0029045150    CÓRDOBA         TIERRALTA
947  0047035030   PUTUMAYO  PUERTO LEGUÍZAMO
948  0056019010      CHOCÓ      BAHÍA SOLANO

[949 rows x 3 columns]

Finalmente recopilamos los dos dataframes de limpieza de coordenadas y de ubicaciones de las 894 estaciones descritas en el dataset inicial. Adicionalmente, asignamos el tipo de dato category a las columnas: 'Dept','City'

In [43]:
dfSta = df_part_f.merge(Ubic_f, how='inner', on='CodSta').sort_values(['Dept','City'])
dfSta[['Dept','City']] = dfSta[['Dept','City']].astype('category')
dfSta = dfSta.sort_values('CodSta').reset_index(drop=True)
dfSta['CodSta'] = dfSta['CodSta'].astype('uint64')
dfSta = dfSta.rename_axis('Sta').reset_index()
dfSta.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 894 entries, 0 to 893
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype   
---  ------   --------------  -----   
 0   Sta      894 non-null    int64   
 1   CodSta   894 non-null    uint64  
 2   NameSta  894 non-null    object  
 3   Lat      894 non-null    float64 
 4   Long     894 non-null    float64 
 5   Dept     894 non-null    category
 6   City     894 non-null    category
dtypes: category(2), float64(2), int64(1), object(1), uint64(1)
memory usage: 59.0+ KB


In [44]:
path_file = r"C:\Users\danie\OneDrive\Documents\Bootcamp_Proyecto\stations.parquet"
dfSta.to_parquet(path = path_file)

#### Registros de precipitaciones

Teniendo en cuenta que contamos con un dataframe 'dfSta' que define claramente las coordenadas y ubicación de cada estación. Conviene obtener de manera similar un solo archivo que recopile los registros de las precipitaciones de todas las estaciones. Para conseguir esto, vamos a concatenar los diferentes archivos parquet y vamos a almacenar en un dataframe solo las columnas correspondientes al código de la estación, el código del sensor, la fecha de observación y el valor de la medida.

In [45]:
path = r'C:\\Users\\danie\\OneDrive\\Documents\\Bootcamp_Proyecto\\CleanDatasets' 
namesFiles = os.listdir(path)

En la siguiente celda importamos y concatenamos los diferentes archivos parquet, adicionalmente cambiamos las unidades de la columna 'Rain' a ${\mu m}$

In [46]:
df_rain = pd.DataFrame()
for i in namesFiles:
    path_file = path + '\\' + i
    df_rain_aux = pd.read_parquet(
                                path_file,
                                columns=['CodigoEstacion','CodigoSensor','FechaObservacion','ValorObservado']
                            )
    df_rain_aux = df_rain_aux.rename({'CodigoEstacion':'CodSta',       
                            'CodigoSensor':'CodSen',   
                            'FechaObservacion':'Date',
                            'ValorObservado':'Rain'
                            }, 
                            axis=1)
    df_rain_aux['Rain'] = df_rain_aux['Rain']*1000
    df_rain_aux['Date'] = pd.to_datetime(df_rain_aux['Date'])
    df_rain_aux['CodSta'] = df_rain_aux['CodSta'].astype('uint64')
    df_rain_aux['CodSen'] = df_rain_aux['CodSen'].astype('uint16')
    df_rain_aux['Rain'] = df_rain_aux['Rain'].astype('uint16')
    df_rain = pd.concat([df_rain,df_rain_aux], ignore_index=True)


df_rain.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 199781152 entries, 0 to 199781151
Data columns (total 4 columns):
 #   Column  Dtype         
---  ------  -----         
 0   CodSta  uint64        
 1   CodSen  uint16        
 2   Date    datetime64[ns]
 3   Rain    uint16        
dtypes: datetime64[ns](1), uint16(2), uint64(1)
memory usage: 3.7 GB


Ajustamos algunos tipos de dato con el objetivo de reducir el tamaño del dataframe y evitar un consumo excesivo de memoria. Sin embargo, es posible realizar un ajuste mas minucioso a cada columna para lograr reducir aún mas el tamaño.

   - La columna CodSta describe 894 estaciones, por tanto podríamos emplear una codificación diferente para optimizar el uso de memoria.
      - Usaremos el índice del dataframe que almacena la ubicación de las estaciones en lugar del código de la estación y le llamaremos a dicha columna Sta.
   - La columna CodSen describe solamente 2 valores diferentes a lo largo del dataframe, de manera que se puede cambiar de manera similar la codificación.
      - Para el Código de sensor 240 usaremos 0 y para el Código de sensor 257 usaremos 1.

In [47]:
path_file = r'C:\Users\danie\OneDrive\Documents\Bootcamp_Proyecto\stations.parquet' 
dfSta = pd.read_parquet(path_file)
dfSta_cod = dfSta['CodSta'].reset_index()
dfSta_cod.rename(columns={'index':'Sta'}, inplace=True)

df_rain = df_rain.merge(dfSta_cod, how='left', on='CodSta')
df_rain = df_rain.drop('CodSta', axis=1)

df_rain.loc[df_rain['CodSen']==240,'CodSen'] = 0
df_rain.loc[df_rain['CodSen']==257,'CodSen'] = 1
df_rain

CodSen                Date  Rain  Sta
0               0 2019-06-10 03:10:00     0  659
1               0 2017-11-12 08:30:00     0  650
2               0 2017-04-26 09:10:00     0  658
3               0 2019-03-03 06:50:00     0  658
4               0 2019-08-03 11:00:00     0  658
...           ...                 ...   ...  ...
199781147       0 2024-10-15 06:20:00     0  879
199781148       0 2024-10-15 10:20:00     0  879
199781149       0 2024-10-15 10:50:00     0  879
199781150       0 2024-10-15 06:50:00     0  879
199781151       0 2024-10-15 07:40:00     0  879

[199781152 rows x 4 columns]

In [48]:
df_rain['CodSen'] = df_rain['CodSen'].astype('uint8')
df_rain['Sta'] = df_rain['Sta'].astype('uint16')
df_rain = df_rain.reindex(['Sta','CodSen','Date','Rain'], axis=1)
df_rain = df_rain.sort_values(['Sta','Date'])
df_rain.info()

<class 'pandas.core.frame.DataFrame'>
Index: 199781152 entries, 84275394 to 196782247
Data columns (total 4 columns):
 #   Column  Dtype         
---  ------  -----         
 0   Sta     uint16        
 1   CodSen  uint8         
 2   Date    datetime64[ns]
 3   Rain    uint16        
dtypes: datetime64[ns](1), uint16(2), uint8(1)
memory usage: 3.9 GB


In [49]:
df_rain

Sta  CodSen                Date  Rain
84275394     0       0 2018-05-04 01:00:00     0
85171743     0       0 2018-05-04 01:10:00     0
84822562     0       0 2018-05-04 01:20:00     0
85194375     0       0 2018-05-04 01:30:00     0
84410973     0       0 2018-05-04 01:40:00     0
...        ...     ...                 ...   ...
196783612  893       0 2020-08-27 11:20:00     0
196781914  893       0 2020-08-27 11:30:00     0
196783171  893       0 2020-08-27 11:40:00     0
196785212  893       0 2020-08-27 11:50:00     0
196782247  893       0 2020-08-27 12:00:00     0

[199781152 rows x 4 columns]

In [50]:
path_file = r"C:\Users\danie\OneDrive\Documents\Bootcamp_Proyecto\rain.parquet"
df_rain.to_parquet(path = path_file)